// testcafe/lib/test-run/phase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var TestRunPhase;
(function (TestRunPhase) {
    TestRunPhase["initial"] = "initial";
    TestRunPhase["inTestRunBeforeHook"] = "inTestRunBeforeHook";
    TestRunPhase["inFixtureBeforeHook"] = "inFixtureBeforeHook";
    TestRunPhase["inFixtureBeforeEachHook"] = "inFixtureBeforeEachHook";
    TestRunPhase["inTestBeforeHook"] = "inTestBeforeHook";
    TestRunPhase["inTest"] = "inTest";
    TestRunPhase["inTestAfterHook"] = "inTestAfterHook";
    TestRunPhase["inFixtureAfterEachHook"] = "inFixtureAfterEachHook";
    TestRunPhase["inFixtureAfterHook"] = "inFixtureAfterHook";
    TestRunPhase["inTestRunAfterHook"] = "inTestRunAfterHook";
    TestRunPhase["inRoleInitializer"] = "inRoleInitializer";
    TestRunPhase["inBookmarkRestore"] = "inBookmarkRestore";
    TestRunPhase["pendingFinalization"] = "pendingFinalization";
})(TestRunPhase || (TestRunPhase = {}));
exports.default = TestRunPhase;
module.exports = exports.default;
