// testcafe/lib/shared/actions/automations/settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ACTION_STEP_DELAY_DEFAULT = 10;
const MAX_MOUSE_ACTION_STEP_DELAY = 400;
const MAX_KEY_ACTION_STEP_DELAY = 200;
// We use an exponential function to calculate the cursor
// speed according to general test speed
// cursorSpeed = (maxSpeed * k) ^ speed / k
const MAX_CURSOR_SPEED = 100; // pixels/ms
const MAX_DRAGGING_SPEED = 4; // pixels/ms
const CURSOR_FACTOR = 4;
class AutomationSettings {
    constructor(speed) {
        this._speedFactor = speed || 1;
    }
    get mouseActionStepDelay() {
        return this._speedFactor === 1 ? ACTION_STEP_DELAY_DEFAULT : (1 - this._speedFactor) * MAX_MOUSE_ACTION_STEP_DELAY;
    }
    get keyActionStepDelay() {
        return this._speedFactor === 1 ? ACTION_STEP_DELAY_DEFAULT : (1 - this._speedFactor) * MAX_KEY_ACTION_STEP_DELAY;
    }
    get cursorSpeed() {
        return Math.pow(MAX_CURSOR_SPEED * CURSOR_FACTOR, this._speedFactor) / CURSOR_FACTOR;
    }
    get draggingSpeed() {
        return Math.pow(MAX_DRAGGING_SPEED * CURSOR_FACTOR, this._speedFactor) / CURSOR_FACTOR;
    }
}
exports.default = AutomationSettings;
module.exports = exports.default;
