// testcafe/lib/client/automation/playback/press/get-key-info.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const is_letter_1 = __importDefault(require("../../utils/is-letter"));
// @ts-ignore
const testcafe_core_1 = require("../../deps/testcafe-core");
const get_key_identifier_1 = __importDefault(require("../../utils/get-key-identifier"));
const get_key_code_1 = __importDefault(require("../../utils/get-key-code"));
function isNewLineKey(key) {
    return testcafe_core_1.KEY_MAPS.newLineKeys.indexOf(key) > -1;
}
function default_1(key, eventKeyProperty) {
    const sanitizedKey = (0, testcafe_core_1.getSanitizedKey)(key);
    const isChar = key.length === 1 || key === 'space';
    const isNewLine = isNewLineKey(key);
    const modifierKeyCode = testcafe_core_1.KEY_MAPS.modifiers[sanitizedKey];
    const specialKeyCode = testcafe_core_1.KEY_MAPS.specialKeys[sanitizedKey];
    eventKeyProperty = eventKeyProperty || key;
    return {
        sanitizedKey,
        isChar,
        modifierKeyCode,
        specialKeyCode,
        isNewLine,
        isLetter: (0, is_letter_1.default)(key),
        keyIdentifierProperty: (0, get_key_identifier_1.default)(eventKeyProperty),
        keyProperty: testcafe_core_1.KEY_MAPS.keyProperty[eventKeyProperty] || eventKeyProperty,
        keyCode: getResultKeyCode(key, isChar, sanitizedKey, modifierKeyCode, specialKeyCode),
    };
}
exports.default = default_1;
function getResultKeyCode(key, isChar, sanitizedKey, modifierKeyCode, specialKeyCode) {
    let keyCode = null;
    if (isChar && key !== 'space')
        keyCode = (0, get_key_code_1.default)(sanitizedKey);
    else if (modifierKeyCode)
        keyCode = modifierKeyCode;
    else if (specialKeyCode)
        keyCode = specialKeyCode;
    return keyCode;
}
