// testcafe/lib/utils/check-is-vm.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkIsVM = void 0;
const os_1 = __importDefault(require("os"));
const promisified_functions_1 = require("./promisified-functions");
const PLATFORM = {
    WIN: 'win32',
    LINUX: 'linux',
    MAC: 'darwin',
};
const VM_REGEX = /virtual|vmware|hyperv|wsl|hyper-v|microsoft|parallels|qemu/gi;
const NOT_FOUND_REGEX_LINUX = /systemd-detect-virt: not found/ig;
const VM_BIOS = '0';
const COMMANDS = {
    WINDOWS_BIOS_NUMBER: 'WMIC BIOS GET SERIALNUMBER',
    WINDOWS_MODEL: 'WMIC COMPUTERSYSTEM GET MODEL',
    WINDOWS_MANUFACTURER: 'WMIC COMPUTERSYSTEM GET MANUFACTURER',
    LINUX_DETECT_VM: 'systemd-detect-virt',
    MAC_MANUFACTURER: 'ioreg -l | grep -e Manufacturer -e \'Vendor Name\'',
};
async function getCommandOutput(command) {
    const { stdout } = await (0, promisified_functions_1.exec)(command);
    return stdout;
}
async function isLinuxVM() {
    let isVM = false;
    try {
        await (0, promisified_functions_1.exec)(COMMANDS.LINUX_DETECT_VM);
        isVM = true;
    }
    catch (error) {
        isVM = NOT_FOUND_REGEX_LINUX.test(error.stderr);
    }
    return isVM;
}
async function isWinVM() {
    const biosNumberOutput = await getCommandOutput(COMMANDS.WINDOWS_BIOS_NUMBER);
    const modelOutput = await getCommandOutput(COMMANDS.WINDOWS_MODEL);
    const manufacturerOutput = await getCommandOutput(COMMANDS.WINDOWS_MANUFACTURER);
    return biosNumberOutput === VM_BIOS || VM_REGEX.test(modelOutput) || VM_REGEX.test(manufacturerOutput);
}
async function isMacVM() {
    return VM_REGEX.test(await getCommandOutput(COMMANDS.MAC_MANUFACTURER));
}
async function checkIsVM() {
    switch (os_1.default.platform()) {
        case PLATFORM.LINUX: return await isLinuxVM();
        case PLATFORM.WIN: return await isWinVM();
        case PLATFORM.MAC: return await isMacVM();
        default:
            return false;
    }
}
exports.checkIsVM = checkIsVM;
