// testcafe/lib/utils/endpoint-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getIPAddress = exports.isMyHostname = exports.getFreePort = exports.isFreePort = void 0;
const net_1 = require("net");
const address_1 = require("address");
function createServerOnFreePort() {
    return new Promise(resolve => {
        const server = (0, net_1.createServer)();
        server.once('listening', () => {
            resolve(server);
        });
        server.listen(0);
    });
}
function closeServers(servers) {
    return Promise.all(servers.map((server) => {
        return new Promise(resolve => {
            server.once('close', resolve);
            server.close();
        });
    }));
}
function checkAvailability(port, hostname) {
    return new Promise(resolve => {
        const server = (0, net_1.createServer)();
        server.once('error', () => {
            resolve(false);
        });
        server.once('listening', () => {
            server.once('close', () => {
                resolve(true);
            });
            server.close();
        });
        server.listen(port, hostname);
    });
}
function isFreePort(port) {
    return checkAvailability(port);
}
exports.isFreePort = isFreePort;
function getFreePort() {
    return getFreePorts(1).then(ports => {
        return ports[0];
    });
}
exports.getFreePort = getFreePort;
function getFreePorts(count) {
    const serverPromises = [];
    let ports = [];
    // NOTE: Sequentially collect listening
    // servers to avoid interference.
    for (let i = 0; i < count; i++)
        serverPromises.push(createServerOnFreePort());
    return Promise.all(serverPromises)
        .then(servers => {
        ports = servers.map((server) => {
            return server.address().port;
        });
        return servers;
    })
        .then(closeServers)
        .then(() => {
        return ports;
    });
}
function isMyHostname(hostname) {
    return getFreePort()
        .then(port => {
        return checkAvailability(port, hostname);
    });
}
exports.isMyHostname = isMyHostname;
function getIPAddress() {
    return (0, address_1.ip)() || (0, address_1.ipv6)();
}
exports.getIPAddress = getIPAddress;
