// testcafe/lib/utils/get-filter-fn.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lodash_1 = require("lodash");
const FILTERING_OPTIONS = {
    testGrep: 'testGrep',
    fixtureGrep: 'fixtureGrep',
    testMeta: 'testMeta',
    fixtureMeta: 'fixtureMeta',
    test: 'test',
    fixture: 'fixture',
};
function isAllFilteringOptionsAreUndefined(opts) {
    return Object
        .keys(FILTERING_OPTIONS)
        .every(option => opts[option] === void 0);
}
function createFilterFn(opts) {
    return (testName, fixtureName, fixturePath, testMeta, fixtureMeta) => {
        if (opts.test && testName !== opts.test)
            return false;
        if (opts.testGrep && !opts.testGrep.test(testName))
            return false;
        if (opts.fixture && fixtureName !== opts.fixture)
            return false;
        if (opts.fixtureGrep && !opts.fixtureGrep.test(fixtureName))
            return false;
        if (opts.testMeta && !lodash_1.isMatch(testMeta, opts.testMeta))
            return false;
        if (opts.fixtureMeta && !lodash_1.isMatch(fixtureMeta, opts.fixtureMeta))
            return false;
        return true;
    };
}
function default_1(opts) {
    if (typeof opts === 'function')
        return opts;
    const filteringOpts = lodash_1.pick(opts, Object.keys(FILTERING_OPTIONS));
    if (isAllFilteringOptionsAreUndefined(filteringOpts))
        return void 0;
    return Object.assign(createFilterFn(filteringOpts), filteringOpts);
}
exports.default = default_1;
module.exports = exports.default;
