// testcafe/lib/utils/parse-user-agent.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseUserAgent = exports.calculatePrettyUserAgent = void 0;
const bowser_1 = __importDefault(require("bowser"));
const DEFAULT_NAME = 'Other';
const DEFAULT_VERSION = '0.0';
const DEFAULT_PLATFORM_TYPE = DEFAULT_NAME.toLowerCase();
const EMPTY_PARSED_USER_AGENT = bowser_1.default.parse(' ');
const HEADLESS_EDGE = {
    regExp: /HeadlessEdg/i,
    browserName: 'Microsoft Edge',
};
function calculateBrowser(browserDetails) {
    return {
        name: browserDetails.name || DEFAULT_NAME,
        version: browserDetails.version || DEFAULT_VERSION,
    };
}
function calculateOs(parsedOsDetails) {
    const name = parsedOsDetails.name || DEFAULT_NAME;
    let version = DEFAULT_VERSION;
    // NOTE: a 'versionName' property value is more readable in the case of Windows (GH-481):
    // Windows 8.1: os.version: "NT 6.3", os.versionName: "8.1".
    if (name.toLowerCase() === 'windows') {
        if (parsedOsDetails.versionName)
            version = parsedOsDetails.versionName;
    }
    else if (parsedOsDetails.version)
        version = parsedOsDetails.version;
    return { name, version };
}
function calculateEngine(engineDetails) {
    return {
        name: engineDetails.name || DEFAULT_NAME,
        version: engineDetails.version || DEFAULT_VERSION,
    };
}
function calculateParsedUserAgent(userAgent) {
    if (!userAgent)
        return EMPTY_PARSED_USER_AGENT;
    // The 'bowser' module incorrectly determine the headless edge browser.
    // Since this module is abandoned, we are forced to fix it on our side.
    const isHeadlessEdge = HEADLESS_EDGE.regExp.test(userAgent);
    const parsedUserAgent = bowser_1.default.parse(userAgent);
    if (isHeadlessEdge)
        parsedUserAgent.browser.name = HEADLESS_EDGE.browserName;
    return parsedUserAgent;
}
function calculatePrettyUserAgent(browser, os) {
    return `${browser.name} ${browser.version} / ${os.name} ${os.version}`;
}
exports.calculatePrettyUserAgent = calculatePrettyUserAgent;
function parseUserAgent(userAgent = '', osInfo) {
    const parsedUserAgent = calculateParsedUserAgent(userAgent);
    const browser = calculateBrowser(parsedUserAgent.browser);
    const os = osInfo || calculateOs(parsedUserAgent.os);
    const engine = calculateEngine(parsedUserAgent.engine);
    const prettyUserAgent = calculatePrettyUserAgent(browser, os);
    return {
        name: browser.name,
        version: browser.version,
        platform: parsedUserAgent.platform.type || DEFAULT_PLATFORM_TYPE,
        os,
        engine,
        prettyUserAgent: prettyUserAgent,
        userAgent,
    };
}
exports.parseUserAgent = parseUserAgent;
