// testcafe-hammerhead/lib/processing/dom/attributes.js
"use strict";
// -------------------------------------------------------------
// WARNING: this file is used by both the client and the server.
// Do not use any browser or node-specific API!
// -------------------------------------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
exports.ATTRS_WITH_SPECIAL_PROXYING_LOGIC = exports.TARGET_ATTRS = exports.TARGET_ATTR_TAGS = exports.URL_ATTRS = exports.URL_ATTR_TAGS = void 0;
exports.URL_ATTR_TAGS = {
    href: ['a', 'link', 'image', 'area', 'base'],
    src: ['img', 'embed', 'script', 'source', 'video', 'audio', 'input', 'frame', 'iframe'],
    srcset: ['img', 'source'],
    action: ['form'],
    formaction: ['button', 'input'],
    manifest: ['html'],
    data: ['object'],
};
exports.URL_ATTRS = ['href', 'src', 'action', 'formaction', 'manifest', 'data'];
exports.TARGET_ATTR_TAGS = {
    target: ['a', 'form', 'area', 'base'],
    formtarget: ['input', 'button'],
};
exports.TARGET_ATTRS = ['target', 'formtarget'];
exports.ATTRS_WITH_SPECIAL_PROXYING_LOGIC = ['sandbox', 'autocomplete', 'target', 'formtarget', 'style'];
