// testcafe-reporter-dashboard/lib/types/task-done-args.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskDoneArgsSchema = exports.TaskResultSchema = void 0;
const io_ts_types_1 = require("io-ts-types");
const t = __importStar(require("io-ts"));
const common_1 = require("./common");
exports.TaskResultSchema = t.readonly(t.exact(t.type({
    failedCount: t.number,
    passedCount: t.number,
    skippedCount: t.number
})));
exports.TaskDoneArgsSchema = t.readonly(t.exact(t.type({
    endTime: io_ts_types_1.DateFromISOString,
    passed: t.number,
    warnings: t.array(t.string),
    warningsUploadId: t.union([t.string, t.undefined]),
    result: exports.TaskResultSchema,
    buildId: common_1.BuildIdSchema
})));
