// testcafe-reporter-dashboard/lib/types/task-start-args.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskStartArgsSchema = exports.CIInfoSchema = exports.ReportedTestStructureItemSchema = exports.ReportedFixtureItemSchema = exports.ReportedTestItemSchema = void 0;
const t = __importStar(require("io-ts"));
const io_ts_types_1 = require("io-ts-types");
const common_1 = require("./common");
exports.ReportedTestItemSchema = t.readonly(t.exact(t.type({
    id: common_1.ShortIdSchema,
    name: common_1.NameSchema,
    skip: t.boolean
})));
exports.ReportedFixtureItemSchema = t.readonly(t.exact(t.type({
    id: common_1.NameSchema,
    name: common_1.NameSchema,
    tests: t.array(exports.ReportedTestItemSchema)
})));
exports.ReportedTestStructureItemSchema = t.readonly(t.exact(t.type({
    fixture: exports.ReportedFixtureItemSchema
})));
exports.CIInfoSchema = t.readonly(t.exact(t.partial({
    commitSHA: t.string,
    author: t.string,
    branchName: t.string
})));
exports.TaskStartArgsSchema = t.readonly(t.exact(t.type({
    startTime: io_ts_types_1.DateFromISOString,
    userAgents: t.array(t.string),
    testCount: t.number,
    buildId: common_1.BuildIdSchema,
    taskStructure: t.array(exports.ReportedTestStructureItemSchema),
    ciInfo: t.union([t.undefined, exports.CIInfoSchema])
})));
