// testcafe-reporter-dashboard/lib/types/testcafe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommandType = exports.TestPhase = void 0;
var TestPhase;
(function (TestPhase) {
    TestPhase["initial"] = "initial";
    TestPhase["inFixtureBeforeHook"] = "inFixtureBeforeHook";
    TestPhase["inFixtureBeforeEachHook"] = "inFixtureBeforeEachHook";
    TestPhase["inTestBeforeHook"] = "inTestBeforeHook";
    TestPhase["inTest"] = "inTest";
    TestPhase["inTestAfterHook"] = "inTestAfterHook";
    TestPhase["inFixtureAfterEachHook"] = "inFixtureAfterEachHook";
    TestPhase["inFixtureAfterHook"] = "inFixtureAfterHook";
    TestPhase["inRoleInitializer"] = "inRoleInitializer";
    TestPhase["inBookmarkRestore"] = "inBookmarkRestore";
})(TestPhase = exports.TestPhase || (exports.TestPhase = {}));
;
var CommandType;
(function (CommandType) {
    CommandType["click"] = "click";
    CommandType["rightClick"] = "right-click";
    CommandType["doubleClick"] = "double-click";
    CommandType["drag"] = "drag";
    CommandType["dragToElement"] = "drag-to-element";
    CommandType["hover"] = "hover";
    CommandType["typeText"] = "type-text";
    CommandType["selectText"] = "select-text";
    CommandType["selectTextAreaContent"] = "select-text-area-content";
    CommandType["selectEditableContent"] = "select-editable-content";
    CommandType["pressKey"] = "press-key";
    CommandType["wait"] = "wait";
    CommandType["navigateTo"] = "navigate-to";
    CommandType["setFilesToUpload"] = "set-files-to-upload";
    CommandType["clearUpload"] = "clear-upload";
    CommandType["executeClientFunction"] = "execute-client-function";
    CommandType["executeSelector"] = "execute-selector";
    CommandType["takeScreenshot"] = "take-screenshot";
    CommandType["takeElementScreenshot"] = "take-element-screenshot";
    CommandType["takeScreenshotOnFail"] = "take-screenshot-on-fail";
    CommandType["prepareBrowserManipulation"] = "prepare-browser-manipulation";
    CommandType["showAssertionRetriesStatus"] = "show-assertion-retries-status";
    CommandType["hideAssertionRetriesStatus"] = "hide-assertion-retries-status";
    CommandType["setBreakpoint"] = "set-breakpoint";
    CommandType["resizeWindow"] = "resize-window";
    CommandType["resizeWindowToFitDevice"] = "resize-window-to-fit-device";
    CommandType["maximizeWindow"] = "maximize-window";
    CommandType["switchToIframe"] = "switch-to-iframe";
    CommandType["switchToMainWindow"] = "switch-to-main-window";
    CommandType["setNativeDialogHandler"] = "set-native-dialog-handler";
    CommandType["getNativeDialogHistory"] = "get-native-dialog-history";
    CommandType["getBrowserConsoleMessages"] = "get-browser-console-messages";
    CommandType["setTestSpeed"] = "set-test-speed";
    CommandType["setPageLoadTimeout"] = "set-page-load-timeout";
    CommandType["debug"] = "debug";
    CommandType["assertion"] = "assertion";
    CommandType["useRole"] = "useRole";
    CommandType["testDone"] = "test-done";
    CommandType["backupStorages"] = "backup-storages";
    CommandType["executeExpression"] = "execute-expression";
    CommandType["executeAsyncExpression"] = "execute-async-expression";
    CommandType["unlockPage"] = "unlock-page";
    CommandType["recorder"] = "recorder";
})(CommandType = exports.CommandType || (exports.CommandType = {}));
;
