// testcafe/lib/native-automation/request-pipeline/safe-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFailRequest = exports.safeContinueRequest = exports.safeFulfillRequest = exports.safeContinueResponse = exports.connectionResetGuard = void 0;
const debug_loggers_1 = require("../../utils/debug-loggers");
const cdp_1 = require("../utils/cdp");
const IGNORED_ERROR_CODES = {
    // In some cases (a request was aborted, any page that initiated the request doesn't exist, etc.)
    // Chrome Debug Protocol doesn't allow to continue request pipeline
    // and raises the "Invalid InterceptionId" error.
    INVALID_INTERCEPTION_ID: -32602,
    // The "Session not found" error can occur in iframes for unclear reasons.
    SESSION_WITH_GIVEN_ID_NOT_FOUND: -32001,
};
//The "WebSocket connection closed" error occurs on closeWindow in multiple windows mode
const IGNORED_ERROR_MESSAGES = ['WebSocket connection closed'];
async function connectionResetGuard(handleRequestFn, handleErrorFn) {
    var _a;
    try {
        await handleRequestFn();
    }
    catch (err) {
        if (Object.values(IGNORED_ERROR_CODES).includes((_a = err === null || err === void 0 ? void 0 : err.response) === null || _a === void 0 ? void 0 : _a.code) || IGNORED_ERROR_MESSAGES.includes(err.message))
            return;
        handleErrorFn(err);
        throw err;
    }
}
exports.connectionResetGuard = connectionResetGuard;
async function safeContinueResponse(client, data, sessionId) {
    const isPausedEvent = (0, cdp_1.isRequestPausedEvent)(data);
    await connectionResetGuard(async () => {
        const param = isPausedEvent
            ? { requestId: data.requestId }
            : data;
        // @ts-ignore
        await client.Fetch.continueResponse(param, sessionId);
    }, err => {
        const formatter = isPausedEvent ? '%r' : '%s';
        (0, debug_loggers_1.requestPipelineLogger)(`Fetch.continueResponse. Unhandled error %s during processing ${formatter}`, err, data);
    });
}
exports.safeContinueResponse = safeContinueResponse;
async function safeFulfillRequest(client, fulfillInfo, sessionId) {
    await connectionResetGuard(async () => {
        // @ts-ignore
        await client.Fetch.fulfillRequest(fulfillInfo, sessionId);
    }, err => {
        (0, debug_loggers_1.requestPipelineLogger)(`Fetch.fulfillRequest. Unhandled error %s during processing %s`, err, fulfillInfo.requestId);
    });
}
exports.safeFulfillRequest = safeFulfillRequest;
async function safeContinueRequest(client, event, sessionId, continueRequestArgs) {
    const { postData, method, url, headers } = continueRequestArgs || {};
    await connectionResetGuard(async () => {
        // @ts-ignore
        await client.Fetch.continueRequest({ requestId: event.requestId, postData, method, url, headers }, sessionId);
    }, err => {
        (0, debug_loggers_1.requestPipelineLogger)(`Fetch.continueRequest. Unhandled error %s during processing %r`, err, event);
    });
}
exports.safeContinueRequest = safeContinueRequest;
async function safeFailRequest(client, event, errorReason = 'Aborted') {
    await connectionResetGuard(async () => {
        await client.Fetch.failRequest({
            requestId: event.requestId,
            errorReason,
        });
    }, err => {
        (0, debug_loggers_1.requestPipelineLogger)(`Fetch.failRequest. Unhandled error %s during processing %s`, err, event.requestId);
    });
}
exports.safeFailRequest = safeFailRequest;
