// testcafe-reporter-dashboard/lib/report-commands-factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const internal_1 = require("./types/internal/");
function reportCommandsFactory(reportId, transport) {
    async function sendReportCommand(type, payload) {
        return transport.sendResolveCommand({
            aggregateId: reportId,
            aggregateName: internal_1.AggregateNames.Run,
            type,
            payload
        });
    }
    return {
        sendTaskStartCommand(payload) {
            return sendReportCommand(internal_1.AggregateCommandType.reportTaskStart, payload);
        },
        sendTestStartCommand(payload) {
            return sendReportCommand(internal_1.AggregateCommandType.reportTestStart, payload);
        },
        sendTestDoneCommand(payload) {
            return sendReportCommand(internal_1.AggregateCommandType.reportTestDone, payload);
        },
        sendTaskDoneCommand(payload) {
            return sendReportCommand(internal_1.AggregateCommandType.reportTaskDone, payload);
        },
        sendReportWarningsCommand(payload) {
            return sendReportCommand(internal_1.AggregateCommandType.reportWarnings, payload);
        }
    };
}
exports.default = reportCommandsFactory;
;
