// testcafe-hammerhead/lib/processing/script/transformers/static-import.js
"use strict";
// -------------------------------------------------------------
// WARNING: this file is used by both the client and the server.
// Do not use any browser or node-specific API!
// -------------------------------------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
const node_builder_1 = require("../node-builder");
const esotope_hammerhead_1 = require("esotope-hammerhead");
// Transform:
// import something from 'url';   -->   import something from 'processed-url';
// export * from 'url';   -->   export * from 'processed-url';
// export { x as y } from 'url';   -->   export { x as y } from 'processed-url';
const transformer = {
    name: 'static-import',
    nodeReplacementRequireTransform: false,
    nodeTypes: esotope_hammerhead_1.Syntax.Literal,
    condition: (node, parent) => !!parent && (parent.type === esotope_hammerhead_1.Syntax.ImportDeclaration ||
        parent.type === esotope_hammerhead_1.Syntax.ExportAllDeclaration ||
        parent.type === esotope_hammerhead_1.Syntax.ExportNamedDeclaration) && parent.source === node,
    run: node => transformer.resolver ? (0, node_builder_1.getProxyUrlLiteral)(node, transformer.resolver) : null,
};
exports.default = transformer;module.exports = exports.default;
