// testcafe/lib/browser/provider/built-in/dedicated/chrome/runtime-info.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const endpoint_utils_1 = require("endpoint-utils");
const create_temp_profile_1 = __importDefault(require("./create-temp-profile"));
const is_docker_1 = __importDefault(require("is-docker"));
class ChromeRuntimeInfo {
    constructor(config) {
        this.config = config;
        this.tempProfileDir = null;
        this.cdpPort = this.config.cdpPort;
        this.inDocker = is_docker_1.default();
    }
    async createTempProfile(proxyHostName, disableMultipleWindows) {
        return await create_temp_profile_1.default(proxyHostName, disableMultipleWindows);
    }
    static async create(proxyHostName, config, disableMultipleWindows) {
        const runtimeInfo = new this(config);
        if (!runtimeInfo.config.userProfile)
            runtimeInfo.tempProfileDir = await runtimeInfo.createTempProfile(proxyHostName, disableMultipleWindows);
        if (!runtimeInfo.cdpPort && !runtimeInfo.config.userProfile)
            runtimeInfo.cdpPort = await endpoint_utils_1.getFreePort();
        return runtimeInfo;
    }
}
exports.default = ChromeRuntimeInfo;
module.exports = exports.default;
