// testcafe/lib/runner/test-run-hook-controller.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const process_test_fn_error_1 = __importDefault(require("../errors/process-test-fn-error"));
const phase_1 = __importDefault(require("../test-run/phase"));
const type_assertions_1 = require("../errors/runtime/type-assertions");
class TestRunHookController {
    constructor(tests, hook) {
        this.beforeFn = hook === null || hook === void 0 ? void 0 : hook.before;
        this.afterFn = hook === null || hook === void 0 ? void 0 : hook.after;
        this.started = false;
        this.runningBeforeHook = false;
        this.beforeHookErr = null;
        this.pendingTestRunCount = tests.length;
        this.testRunCtx = Object.create(null);
        this._assertHooks();
    }
    _assertHooks() {
        if (this.beforeFn)
            (0, type_assertions_1.assertType)(type_assertions_1.is.function, 'globalBefore', 'The testRun.globalBefore hook', this.beforeFn);
        if (this.afterFn)
            (0, type_assertions_1.assertType)(type_assertions_1.is.function, 'globalAfter', 'The testRun.globalAfter hook', this.afterFn);
    }
    isTestBlocked() {
        return this.runningBeforeHook;
    }
    async runTestRunBeforeHookIfNecessary(testRun) {
        const shouldRunBeforeHook = !this.started && this.beforeFn;
        this.started = true;
        if (shouldRunBeforeHook) {
            this.runningBeforeHook = true;
            try {
                await this.beforeFn(this.testRunCtx);
            }
            catch (err) {
                this.beforeHookErr = (0, process_test_fn_error_1.default)(err);
            }
            this.runningBeforeHook = false;
        }
        // NOTE: fail all tests if testRun.before hook has error
        if (this.beforeHookErr) {
            testRun.phase = phase_1.default.inTestRunBeforeHook;
            testRun.addError(this.beforeHookErr);
            return false;
        }
        testRun.testRunCtx = this.testRunCtx;
        return true;
    }
    async runTestRunAfterHookIfNecessary(testRun) {
        this.pendingTestRunCount--;
        if (this.pendingTestRunCount === 0 && this.afterFn) {
            testRun.phase = phase_1.default.inTestRunAfterHook;
            try {
                await this.afterFn(this.testRunCtx);
            }
            catch (err) {
                testRun.addError((0, process_test_fn_error_1.default)(err));
            }
        }
    }
}
exports.default = TestRunHookController;
module.exports = exports.default;
