// testcafe/lib/live/keyboard-observer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const process_1 = __importDefault(require("process"));
const readline_1 = require("readline");
const lodash_1 = require("lodash");
const LOCK_KEY_PRESS_TIMEOUT = 1000;
let instance;
class LiveModeKeyboardEventObserver {
    constructor() {
        this.controllers = [];
        this.lockKeyPress = false;
        if (!instance) {
            this._listenKeyEvents();
            instance = this;
        }
        return instance;
    }
    push(controller) {
        this.controllers.push(controller);
        if (process_1.default.stdin.isTTY)
            this.setRawMode(true);
    }
    remove(controller) {
        (0, lodash_1.pull)(this.controllers, controller);
        if (!this.controllers.length)
            this.setRawMode(false);
    }
    _listenKeyEvents() {
        (0, readline_1.emitKeypressEvents)(process_1.default.stdin);
        process_1.default.stdin.on('keypress', this._onKeyPress.bind(this));
    }
    setRawMode(value) {
        if (process_1.default.stdin.setRawMode !== void 0)
            process_1.default.stdin.setRawMode(value);
    }
    _onKeyPress(string, key) {
        if (this.lockKeyPress)
            return;
        this.lockKeyPress = true;
        setTimeout(() => {
            this.lockKeyPress = false;
        }, LOCK_KEY_PRESS_TIMEOUT);
        if (key && key.ctrl) {
            switch (key.name) {
                case 's':
                    this._stop();
                    return;
                case 'r':
                    this._restart();
                    return;
                case 'c':
                    this._exit();
                    return;
                case 'w':
                    this._toggleWatching();
                    return;
            }
        }
    }
    _stop() {
        this.controllers.forEach(c => c.stop());
    }
    _restart() {
        this.controllers.forEach(c => c.restart());
    }
    _exit() {
        this.controllers.forEach(c => c.exit());
    }
    _toggleWatching() {
        this.controllers.forEach(c => c.toggleWatching());
    }
}
exports.default = LiveModeKeyboardEventObserver;
module.exports = exports.default;
