// testcafe/lib/live/test-run.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestRunCtorFactory = void 0;
const test_run_1 = __importDefault(require("../test-run"));
const test_run_state_1 = __importDefault(require("./test-run-state"));
const type_1 = __importDefault(require("../test-run/commands/type"));
const service_1 = require("../test-run/commands/service");
const TEST_RUN_ABORTED_MESSAGE = 'The test run has been aborted.';
const TestRunCtorFactory = function (callbacks) {
    const { created, done, readyToNext } = callbacks;
    return class LiveModeTestRun extends test_run_1.default {
        constructor(test, browserConnection, screenshotCapturer, warningLog, opts) {
            super(test, browserConnection, screenshotCapturer, warningLog, opts);
            created(this, test);
            this.state = test_run_state_1.default.created;
            this.finish = null;
            this.stopping = false;
            this.isInRoleInitializing = false;
            this.stopped = false;
        }
        stop() {
            this.stopped = true;
        }
        _useRole(...args) {
            this.isInRoleInitializing = true;
            return super._useRole.apply(this, args)
                .then(res => {
                this.isInRoleInitializing = false;
                return res;
            })
                .catch(err => {
                this.isInRoleInitializing = false;
                throw err;
            });
        }
        _internalExecuteCommand(commandToExec, callsite, forced) {
            // NOTE: don't close the page and the session when the last test in the queue is done
            if (commandToExec.type === type_1.default.testDone && !forced) {
                done(this, this.stopped)
                    .then(() => this._internalExecuteCommand(commandToExec, callsite, true))
                    .then(() => readyToNext(this));
                this._internalExecuteCommand(new service_1.UnlockPageCommand(), null);
                return Promise.resolve();
            }
            if (this.stopped && !this.stopping &&
                !this.isInRoleInitializing) {
                this.stopping = true;
                return Promise.reject(new Error(TEST_RUN_ABORTED_MESSAGE));
            }
            return super._internalExecuteCommand(commandToExec, callsite);
        }
    };
};
exports.TestRunCtorFactory = TestRunCtorFactory;
