// abcolor/build/utility/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var isNumeric = exports.isNumeric = function isNumeric(n) {
  return !isNaN(parseFloat(n)) && isFinite(n);
};

var hashChar = exports.hashChar = function hashChar(str) {
  return str.charAt(0) === '#' ? str.slice(1) : str;
};

var aHex = exports.aHex = function aHex(str) {
  return (/^#?([0-9a-f]{3}){1,2}$/i.test(str)
  );
};

var aTri = exports.aTri = function aTri(arr) {
  return Array.isArray(arr) && arr.length === 3 && arr.every(function (val) {
    return isNumeric(val);
  });
};

var lc = exports.lc = function lc(str) {
  return str.toLowerCase();
};

var formatRgb = exports.formatRgb = function formatRgb(arr) {
  var percent = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
  return aTri(arr) && (percent === true ? 'rgb(' + arr.map(function (val) {
    return val + '%';
  }) + ')' : 'rgb(' + arr.join() + ')');
};

var formatHsl = exports.formatHsl = function formatHsl(arr) {
  return aTri(arr) && 'hsl(' + arr.map(function (val, i) {
    return i > 0 ? val + '%' : val;
  }) + ')';
};
