// testcafe/lib/compiler/test-file/formats/coffeescript/compiler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const coffeescript_1 = __importDefault(require("coffeescript"));
const load_libs_1 = __importDefault(require("../../../babel/load-libs"));
const compiler_js_1 = __importDefault(require("../es-next/compiler.js"));
const extensions_1 = __importDefault(require("../extensions"));
const FIXTURE_RE = /(^|;|\s+)fixture\s*(\.|\(|'|")/;
const TEST_RE = /(^|;|\s+)test\s*/;
class CoffeeScriptTestFileCompiler extends compiler_js_1.default {
    _hasTests(code) {
        return FIXTURE_RE.test(code) && TEST_RE.test(code);
    }
    _compileCode(code, filename) {
        if (this.cache[filename])
            return this.cache[filename];
        const transpiled = coffeescript_1.default.compile(code, {
            filename,
            bare: true,
            sourceMap: true,
            inlineMap: true,
            header: false,
        });
        const { babel } = (0, load_libs_1.default)();
        const babelOptions = compiler_js_1.default.getBabelOptions(filename, code, this);
        const compiled = babel.transform(transpiled.js, babelOptions);
        this.cache[filename] = compiled.code;
        return compiled.code;
    }
    _getRequireCompilers() {
        return { [extensions_1.default.coffee]: (code, filename) => this._compileCode(code, filename) };
    }
    getSupportedExtension() {
        return extensions_1.default.coffee;
    }
}
exports.default = CoffeeScriptTestFileCompiler;
module.exports = exports.default;
