// testcafe/lib/compiler/test-file/formats/coffeescript/get-test-list.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCoffeeScriptTestListFromCode = exports.getCoffeeScriptTestList = exports.CoffeeScriptTestFileParser = void 0;
const coffeescript_1 = __importDefault(require("coffeescript"));
const core_1 = require("@babel/core");
const compiler_js_1 = __importDefault(require("../es-next/compiler.js"));
const get_test_list_1 = require("../es-next/get-test-list");
class CoffeeScriptTestFileParser extends get_test_list_1.EsNextTestFileParser {
    parse(code) {
        const babelOptions = compiler_js_1.default.getBabelOptions(null, code);
        delete babelOptions.filename;
        babelOptions.ast = true;
        code = coffeescript_1.default.compile(code, {
            bare: true,
            sourceMap: false,
            inlineMap: false,
            header: false,
        });
        const ast = (0, core_1.transform)(code, babelOptions).ast;
        return this.analyze(ast.program.body);
    }
}
exports.CoffeeScriptTestFileParser = CoffeeScriptTestFileParser;
const parser = new CoffeeScriptTestFileParser();
exports.getCoffeeScriptTestList = parser.getTestList.bind(parser);
exports.getCoffeeScriptTestListFromCode = parser.getTestListFromCode.bind(parser);
