// testcafe/lib/browser/provider/built-in/dedicated/firefox/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const argument_parsing_1 = require("../../../utils/argument-parsing");
const AVAILABLE_MODES = ['userProfile', 'headless'];
const configCache = {};
function hasCustomProfile(userArgs) {
    return !!(userArgs.match(/-P\s/) || userArgs.match(/-profile\s/));
}
function parseModes(modesStr, userArgs) {
    const parsed = (0, argument_parsing_1.splitEscaped)(modesStr, ':');
    const path = (0, argument_parsing_1.getPathFromParsedModes)(parsed, AVAILABLE_MODES);
    const detectedModes = (0, argument_parsing_1.getModes)(parsed, AVAILABLE_MODES);
    const optionsString = parsed.filter(item => !!item).join(':');
    const options = parsed.length ? (0, argument_parsing_1.splitEscaped)(optionsString, ';') : [];
    return {
        path: path,
        userProfile: detectedModes.userProfile || hasCustomProfile(userArgs),
        headless: detectedModes.headless,
        marionettePort: (0, argument_parsing_1.findMatch)(options, /^marionettePort=(.*)/),
        disableMultiprocessing: (0, argument_parsing_1.isMatchTrue)(options, /^disableMultiprocessing=(.*)/),
    };
}
function getNewConfig(configString) {
    const { userArgs, modesString } = (0, argument_parsing_1.parseConfig)(configString);
    const modes = parseModes(modesString, userArgs);
    return Object.assign({ userArgs }, modes);
}
function default_1(configString) {
    if (!configCache[configString])
        configCache[configString] = getNewConfig(configString);
    return configCache[configString];
}
exports.default = default_1;
module.exports = exports.default;
