// testcafe/lib/browser/provider/built-in/dedicated/firefox/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const os_family_1 = __importDefault(require("os-family"));
const base_1 = __importDefault(require("../base"));
const runtime_info_1 = __importDefault(require("./runtime-info"));
const config_1 = __importDefault(require("./config"));
const local_firefox_1 = require("./local-firefox");
const marionette_client_1 = __importDefault(require("./marionette-client"));
exports.default = Object.assign(Object.assign({}, base_1.default), { getConfig(name) {
        return (0, config_1.default)(name);
    },
    _getBrowserProtocolClient(runtimeInfo) {
        return runtimeInfo.marionetteClient;
    },
    async _createMarionetteClient(runtimeInfo) {
        try {
            const marionetteClient = new marionette_client_1.default(runtimeInfo.marionettePort, runtimeInfo);
            await marionetteClient.connect();
            return marionetteClient;
        }
        catch (e) {
            return null;
        }
    },
    async openBrowser(browserId, pageUrl, config, { disableMultipleWindows }) {
        const runtimeInfo = await (0, runtime_info_1.default)(config);
        runtimeInfo.browserName = this._getBrowserName();
        runtimeInfo.browserId = browserId;
        runtimeInfo.windowDescriptors = {};
        await (0, local_firefox_1.start)(pageUrl, runtimeInfo);
        await this.waitForConnectionReady(runtimeInfo.browserId);
        if (!disableMultipleWindows)
            runtimeInfo.activeWindowId = this.calculateWindowId();
        if (runtimeInfo.marionettePort)
            runtimeInfo.marionetteClient = await this._createMarionetteClient(runtimeInfo);
        this.openedBrowsers[browserId] = runtimeInfo;
    },
    async closeBrowser(browserId) {
        const runtimeInfo = this.openedBrowsers[browserId];
        const { config, marionetteClient } = runtimeInfo;
        if (config.headless)
            await marionetteClient.quit();
        else
            await this.closeLocalBrowser(browserId);
        if (os_family_1.default.mac && !config.headless)
            await (0, local_firefox_1.stop)(runtimeInfo);
        if (runtimeInfo.tempProfileDir)
            await runtimeInfo.tempProfileDir.dispose();
        delete this.openedBrowsers[browserId];
    },
    async resizeWindow(browserId, width, height) {
        const { marionetteClient } = this.openedBrowsers[browserId];
        await marionetteClient.setWindowSize(width, height);
    },
    async getVideoFrameData(browserId) {
        const { marionetteClient } = this.openedBrowsers[browserId];
        return marionetteClient.getScreenshotData();
    },
    async hasCustomActionForBrowser(browserId) {
        const { config, marionetteClient } = this.openedBrowsers[browserId];
        return {
            hasCloseBrowser: true,
            hasTakeScreenshot: !!marionetteClient,
            hasChromelessScreenshots: !!marionetteClient,
            hasGetVideoFrameData: !!marionetteClient,
            hasResizeWindow: !!marionetteClient && config.headless,
            hasMaximizeWindow: !!marionetteClient && config.headless,
            hasCanResizeWindowToDimensions: false,
        };
    } });
module.exports = exports.default;
