// testcafe-browser-tools/lib/aliases.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const binaries_1 = __importDefault(require("./binaries"));
const chromiumCmdArgs = [
    '--disable-infobars',
    '--disable-session-crashed-bubble',
    '--no-first-run',
    '--new-window',
    '--disable-background-networking',
    '--disable-ipc-flooding-protection',
    '--disable-background-timer-throttling',
    '--disable-backgrounding-occluded-windows',
    '--autoplay-policy=no-user-gesture-required'
].join(' ');
const edgeCmdArgs = [
    '--new-window',
    '--disable-background-timer-throttling',
    '--disable-backgrounding-occluded-windows'
].join(' ');
const ALIASES = {
    'ie': {
        nameRe: /iexplore|internet explorer/i,
        cmd: ''
    },
    'firefox': {
        nameRe: /firefox|mozilla/i,
        cmd: `-override "${path_1.default.join(__dirname, '../data/override.ini')}" -new-window`,
        macOpenCmdTemplate: 'open -a "{{{path}}}" {{{pageUrl}}} --args {{{cmd}}}',
        linuxBinaries: ['firefox']
    },
    'chrome-canary': {
        nameRe: /chrome\s*canary/i,
        cmd: chromiumCmdArgs,
        macOpenCmdTemplate: 'open -n -a "{{{path}}}" --args {{{pageUrl}}} {{{cmd}}}',
        linuxBinaries: ['google-chrome-canary']
    },
    'chrome': {
        nameRe: /chrome/i,
        cmd: chromiumCmdArgs,
        macOpenCmdTemplate: 'open -n -a "{{{path}}}" --args {{{pageUrl}}} {{{cmd}}}',
        linuxBinaries: ['google-chrome', 'google-chrome-stable', 'google-chrome-unstable']
    },
    'chromium': {
        nameRe: /chromium/i,
        cmd: chromiumCmdArgs,
        macOpenCmdTemplate: 'open -n -a "{{{path}}}" --args {{{pageUrl}}} {{{cmd}}}',
        linuxBinaries: ['chromium-browser', 'chromium']
    },
    'opera': {
        nameRe: /opera/i,
        cmd: '--new-window',
        macOpenCmdTemplate: 'open -n -a "{{{path}}}" --args {{{pageUrl}}} {{{cmd}}}',
        linuxBinaries: ['opera']
    },
    'safari': {
        nameRe: /safari/i,
        cmd: '',
        path: binaries_1.default.app,
        macOpenCmdTemplate: `open -n -a "{{{path}}}" --args "${binaries_1.default.main}" /dev/null open {{{pageUrl}}} {{{cmd}}}`
    },
    'edge': {
        nameRe: /edge/i,
        cmd: edgeCmdArgs,
        macOpenCmdTemplate: 'open -n -a "{{{path}}}" --args {{{pageUrl}}} {{{cmd}}}',
        linuxBinaries: ['microsoft-edge']
    },
    'edge-legacy': {
        cmd: '',
        winOpenCmdTemplate: 'start microsoft-edge:"{{{pageUrl}}}"'
    }
};
exports.default = ALIASES;
