// testcafe-browser-tools/lib/exports.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.errors = void 0;
var get_installations_1 = require("./api/get-installations");
Object.defineProperty(exports, "getInstallations", { enumerable: true, get: function () { return get_installations_1.default; } });
var find_window_1 = require("./api/find-window");
Object.defineProperty(exports, "findWindow", { enumerable: true, get: function () { return find_window_1.default; } });
var open_1 = require("./api/open");
Object.defineProperty(exports, "open", { enumerable: true, get: function () { return open_1.default; } });
var close_1 = require("./api/close");
Object.defineProperty(exports, "close", { enumerable: true, get: function () { return close_1.default; } });
var resize_1 = require("./api/resize");
Object.defineProperty(exports, "resize", { enumerable: true, get: function () { return resize_1.default; } });
var maximize_1 = require("./api/maximize");
Object.defineProperty(exports, "maximize", { enumerable: true, get: function () { return maximize_1.default; } });
var is_maximized_1 = require("./api/is-maximized");
Object.defineProperty(exports, "isMaximized", { enumerable: true, get: function () { return is_maximized_1.default; } });
var is_valid_device_name_1 = require("./api/is-valid-device-name");
Object.defineProperty(exports, "isValidDeviceName", { enumerable: true, get: function () { return is_valid_device_name_1.default; } });
var screenshot_1 = require("./api/screenshot");
Object.defineProperty(exports, "screenshot", { enumerable: true, get: function () { return screenshot_1.default; } });
var generate_thumbnail_1 = require("./api/generate-thumbnail");
Object.defineProperty(exports, "generateThumbnail", { enumerable: true, get: function () { return generate_thumbnail_1.default; } });
var get_browser_info_1 = require("./api/get-browser-info");
Object.defineProperty(exports, "getBrowserInfo", { enumerable: true, get: function () { return get_browser_info_1.default; } });
var get_viewport_size_1 = require("./api/get-viewport-size");
Object.defineProperty(exports, "getViewportSize", { enumerable: true, get: function () { return get_viewport_size_1.default; } });
var watch_window_1 = require("./api/watch-window");
Object.defineProperty(exports, "watchWindow", { enumerable: true, get: function () { return watch_window_1.default; } });
var bring_to_front_1 = require("./api/bring-to-front");
Object.defineProperty(exports, "bringWindowToFront", { enumerable: true, get: function () { return bring_to_front_1.default; } });
var get_devices_viewport_data_1 = require("./api/get-devices-viewport-data");
Object.defineProperty(exports, "getDevicesViewportData", { enumerable: true, get: function () { return get_devices_viewport_data_1.default; } });
const errors = __importStar(require("./errors"));
exports.errors = errors;
