// testcafe-browser-tools/lib/templates.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UNABLE_TO_OPEN_DISPLAY = exports.UNABLE_TO_ACCESS_SCREEN_RECORDING_API = exports.UNABLE_TO_ACCESS_AUTOMATION_API = exports.NATIVE_BINARY_HAS_FAILED = exports.UNABLE_TO_RUN_BROWSERS = exports.BROWSER_PATH_NOT_SET = exports.BASIC_ERROR_MESSAGE = void 0;
const util_1 = require("util");
const dedent_1 = __importDefault(require("dedent"));
const PROCESS_OUTPUT = ({ output }) => dedent_1.default `
    Process output:
    ${output}
`;
exports.BASIC_ERROR_MESSAGE = data => dedent_1.default `
    An error was thrown.
    Error data:
    ${util_1.inspect(data)}
`;
exports.BROWSER_PATH_NOT_SET = () => dedent_1.default `
    Unable to run the browser. The browser path or command template is not specified.
`;
exports.UNABLE_TO_RUN_BROWSERS = ({ path }) => dedent_1.default `
    Unable to run the browser. The file at ${path} does not exist or is not executable.
`;
exports.NATIVE_BINARY_HAS_FAILED = ({ binary, exitCode, output }) => dedent_1.default `
    The ${binary} process failed with the ${exitCode} exit code.
    ${output ? PROCESS_OUTPUT({ output }) : ''}
`;
exports.UNABLE_TO_ACCESS_AUTOMATION_API = ({ binary }) => dedent_1.default `
    The ${binary} process cannot access the Automation API.
`;
exports.UNABLE_TO_ACCESS_SCREEN_RECORDING_API = ({ binary }) => dedent_1.default `
    The ${binary} process cannot access the Screen Recording API.
`;
exports.UNABLE_TO_OPEN_DISPLAY = ({ binary }) => dedent_1.default `
    The ${binary} process cannot open the display.
`;
