// testcafe-reporter-dashboard/lib/env/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CI_INFO = exports.RESPONSE_TIMEOUT = exports.REQUEST_RETRY_COUNT = exports.TESTCAFE_DASHBOARD_BUILD_ID = exports.TESTCAFE_DASHBOARD_AUTHENTICATION_TOKEN = exports.NO_VIDEO_UPLOAD = exports.NO_SCREENSHOT_UPLOAD = exports.TESTCAFE_DASHBOARD_URL = exports.ENABLE_LOG = void 0;
const utils_1 = require("./utils");
const get_ci_info_1 = require("./get-ci-info");
const { env } = process;
exports.ENABLE_LOG = utils_1.parseBooleanVariable(env.TESTCAFE_DASHBOARD_ENABLE_LOG);
exports.TESTCAFE_DASHBOARD_URL = env.TESTCAFE_DASHBOARD_URL || 'https://dashboard.testcafe.io';
exports.NO_SCREENSHOT_UPLOAD = utils_1.parseBooleanVariable(env.TESTCAFE_DASHBOARD_NO_SCREENSHOT_UPLOAD);
exports.NO_VIDEO_UPLOAD = utils_1.parseBooleanVariable(env.TESTCAFE_DASHBOARD_NO_VIDEO_UPLOAD);
exports.TESTCAFE_DASHBOARD_AUTHENTICATION_TOKEN = env.TESTCAFE_DASHBOARD_TOKEN;
exports.TESTCAFE_DASHBOARD_BUILD_ID = env.TESTCAFE_DASHBOARD_BUILD_ID;
exports.REQUEST_RETRY_COUNT = utils_1.parseNumber(env.TESTCAFE_DASHBOARD_REQUEST_RETRY_COUNT) || 20;
exports.RESPONSE_TIMEOUT = utils_1.parseNumber(env.TESTCAFE_DASHBOARD_RESPONSE_TIMEOUT) || 30 * 1000;
exports.CI_INFO = get_ci_info_1.getCIInfo();
