// testcafe/lib/shared/utils/elements-retriever.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const adapter_1 = require("../adapter");
const node_type_descriptions_1 = __importDefault(require("../utils/node-type-descriptions"));
const errors_1 = require("../../shared/errors");
class ElementsRetriever {
    constructor(globalSelectorTimeout, executeSelectorFn) {
        this._globalSelectorTimeout = globalSelectorTimeout;
        this._ensureElementsStartTime = adapter_1.adapter.nativeMethods.dateNow();
        this._ensureElementsPromise = adapter_1.adapter.PromiseCtor.resolve();
        this._executeSelectorFn = executeSelectorFn;
        this._elements = [];
    }
    push(selector, elementName) {
        this._ensureElementsPromise = this._ensureElementsPromise
            .then(() => {
            return this._executeSelectorFn(selector, {
                invisible: !elementName ? 'ActionElementIsInvisibleError' : {
                    name: 'ActionAdditionalElementIsInvisibleError',
                    firstArg: elementName,
                },
                notFound: !elementName ? 'ActionElementNotFoundError' : {
                    name: 'ActionAdditionalElementNotFoundError',
                    firstArg: elementName,
                },
            }, this._ensureElementsStartTime);
        })
            .then(el => {
            if (!adapter_1.adapter.dom.isDomElement(el)) {
                const nodeType = el.nodeType;
                const nodeTypeStr = node_type_descriptions_1.default[nodeType];
                if (!elementName)
                    throw new errors_1.ActionSelectorMatchesWrongNodeTypeError(nodeTypeStr);
                else
                    throw new errors_1.ActionAdditionalSelectorMatchesWrongNodeTypeError(elementName, nodeTypeStr);
            }
            this._elements.push(el);
        });
    }
    getElements() {
        return this._ensureElementsPromise.then(() => this._elements);
    }
}
exports.default = ElementsRetriever;
module.exports = exports.default;
