// testcafe/lib/shared/utils/values/axis-values.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class AxisValues {
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }
    static create(a) {
        if ('left' in a)
            return new AxisValues(a.left, a.top);
        else if ('right' in a)
            return new AxisValues(a.right, a.bottom);
        return new AxisValues(a.x, a.y);
    }
    add(p) {
        this.x += p.x;
        this.y += p.y;
        return this;
    }
    sub(p) {
        this.x -= p.x;
        this.y -= p.y;
        return this;
    }
    round(fn = Math.round) {
        this.x = fn(this.x);
        this.y = fn(this.y);
        return this;
    }
    eql(p) {
        return this.x === p.x && this.y === p.y;
    }
    mul(n) {
        this.x *= n;
        this.y *= n;
        return this;
    }
    distance(p) {
        return Math.sqrt(Math.pow(this.x - p.x, 2) + Math.pow(this.y - p.y, 2));
    }
}
exports.default = AxisValues;
module.exports = exports.default;
