// testcafe/lib/shared/utils/values/boundary-values.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class BoundaryValues {
    constructor(top = 0, right = 0, bottom = 0, left = 0) {
        this.top = top;
        this.right = right;
        this.bottom = bottom;
        this.left = left;
    }
    static create(v) {
        return new BoundaryValues(v.top, v.right, v.bottom, v.left);
    }
    add(d) {
        this.top += d.top;
        this.right += d.right;
        this.bottom += d.bottom;
        this.left += d.left;
        return this;
    }
    sub(d) {
        if ('top' in d) {
            this.top -= d.top;
            this.left -= d.left;
        }
        this.bottom -= d.bottom;
        this.right -= d.right;
        return this;
    }
    round(leftTopRound = Math.round, rightBottomRound = leftTopRound) {
        this.top = leftTopRound(this.top);
        this.right = rightBottomRound(this.right);
        this.bottom = rightBottomRound(this.bottom);
        this.left = leftTopRound(this.left);
        return this;
    }
    contains(point) {
        return point.x >= this.left && point.x <= this.right && point.y >= this.top && point.y <= this.bottom;
    }
}
exports.default = BoundaryValues;
module.exports = exports.default;
