// testcafe-reporter-dashboard/lib/transport/fetch-response.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class FetchResponse {
    constructor(response, statusText = '', error) {
        this.ok = false;
        this.status = 0;
        if (response) {
            this.ok = response.ok;
            this.status = response.status;
            this.statusText = response.statusText;
        }
        else
            this.statusText = statusText;
        this._error = error;
        this._response = response;
    }
    toString() {
        let result = `${this.status} - ${this.statusText}`;
        if (this._error)
            result += `. ${this._error}`;
        if (this._response)
            result += `. Original response: ${JSON.stringify(this._response)}`;
        return result;
    }
    async json() {
        if (this._response)
            return await this._response.json();
        return null;
    }
    async text() {
        if (this._response)
            return await this._response.text();
        return void 0;
    }
}
exports.default = FetchResponse;
