// testcafe/lib/browser/provider/built-in/dedicated/chrome/cdp-client/automations/click/create-mouse-click-strategy.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const mouse_click_strategy_base_1 = require("../../../../../../../../shared/actions/automations/click/mouse-click-strategy-base");
const clientsManager = __importStar(require("../../clients-manager"));
const BUTTON = 'left';
const CLICK_COUNT = 1;
class CDPMouseClickStategy extends mouse_click_strategy_base_1.MouseClickStrategyBase {
    constructor() {
        super();
    }
    async mousedown(options) {
        var _a, _b, _c, _d;
        const { Input } = clientsManager.getClient();
        await Input.dispatchMouseEvent({
            type: 'mousePressed',
            x: (_b = (_a = options.point) === null || _a === void 0 ? void 0 : _a.x) !== null && _b !== void 0 ? _b : -1,
            y: (_d = (_c = options.point) === null || _c === void 0 ? void 0 : _c.y) !== null && _d !== void 0 ? _d : -1,
            button: BUTTON,
            clickCount: CLICK_COUNT,
        });
    }
    async mouseup(element, options) {
        var _a, _b, _c, _d;
        const { Input } = clientsManager.getClient();
        await Input.dispatchMouseEvent({
            type: 'mouseReleased',
            x: (_b = (_a = options.point) === null || _a === void 0 ? void 0 : _a.x) !== null && _b !== void 0 ? _b : -1,
            y: (_d = (_c = options.point) === null || _c === void 0 ? void 0 : _c.y) !== null && _d !== void 0 ? _d : -1,
            button: BUTTON,
            clickCount: CLICK_COUNT,
        });
        return options;
    }
}
function createMouseClickStrategy() {
    return new CDPMouseClickStategy();
}
exports.default = createMouseClickStrategy;
module.exports = exports.default;
