// testcafe-safe-storage/lib/errors.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultipleSavedDataDetected = exports.SavedDataNotDetected = exports.LoadedDataInvalid = exports.CODES = void 0;
const TEXTS = __importStar(require("./texts.js"));
var CODES;
(function (CODES) {
    CODES[CODES["E001"] = 1] = "E001";
    CODES[CODES["E002"] = 2] = "E002";
    CODES[CODES["E003"] = 3] = "E003";
})(CODES = exports.CODES || (exports.CODES = {}));
class BaseError extends Error {
    constructor(...args) {
        super();
        this.name = new.target.name;
        this.args = args;
    }
    get message() {
        return this.template.format(...this.args);
    }
}
class LoadedDataInvalid extends BaseError {
    constructor() {
        super(...arguments);
        this.code = CODES.E001;
        this.template = TEXTS.LOADED_DATA_INVALID;
    }
}
exports.LoadedDataInvalid = LoadedDataInvalid;
class SavedDataNotDetected extends BaseError {
    constructor() {
        super(...arguments);
        this.code = CODES.E002;
        this.template = TEXTS.SAVED_DATA_NOT_DETECTED;
    }
}
exports.SavedDataNotDetected = SavedDataNotDetected;
class MultipleSavedDataDetected extends BaseError {
    constructor() {
        super(...arguments);
        this.code = CODES.E003;
        this.template = TEXTS.MULTIPLE_SAVED_DATA_DETECTED;
    }
}
exports.MultipleSavedDataDetected = MultipleSavedDataDetected;
