// testcafe/lib/client/automation/utils/key-identifier-maps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    SPECIAL_KEYS: {
        capslock: 'CapsLock',
        delete: 'U+007F',
        end: 'End',
        enter: 'Enter',
        esc: 'U+001B',
        home: 'Home',
        ins: 'Insert',
        pagedown: 'PageDown',
        pageup: 'PageUp',
        space: 'U+0020',
        tab: 'Tab',
        alt: 'Alt',
        ctrl: 'Control',
        meta: 'Meta',
        shift: 'Shift',
    },
    LETTERS: {
        a: 'U+0041',
        b: 'U+0042',
        c: 'U+0043',
        d: 'U+0044',
        e: 'U+0045',
        f: 'U+0046',
        g: 'U+0047',
        h: 'U+0048',
        i: 'U+0049',
        j: 'U+004A',
        k: 'U+004B',
        l: 'U+004C',
        m: 'U+004D',
        n: 'U+004E',
        o: 'U+004F',
        p: 'U+0050',
        q: 'U+0051',
        r: 'U+0052',
        s: 'U+0053',
        t: 'U+0054',
        u: 'U+0055',
        v: 'U+0056',
        w: 'U+0057',
        x: 'U+0058',
        y: 'U+0059',
        z: 'U+005A',
    },
    SYMBOLS: {
        '0': 'U+0030',
        '1': 'U+0031',
        '2': 'U+0032',
        '3': 'U+0033',
        '4': 'U+0034',
        '5': 'U+0035',
        '6': 'U+0036',
        '7': 'U+0037',
        '8': 'U+0038',
        '9': 'U+0039',
        ' ': 'U+0020',
        '!': 'U+0021',
        '@': 'U+0040',
        '#': 'U+0023',
        '$': 'U+0024',
        '%': 'U+0025',
        '^': 'U+005E',
        '*': 'U+002A',
        '(': 'U+0028',
        ')': 'U+0029',
        '_': 'U+005F',
        '|': 'U+007C',
        '\\': 'U+005C',
        '/': 'U+002F',
        '?': 'U+003F',
        '.': 'U+002E',
        ',': 'U+002C',
        '<': 'U+003C',
        '>': 'U+003E',
        '[': 'U+005B',
        ']': 'U+005D',
        '{': 'U+007B',
        '}': 'U+007D',
        '§': 'U+00A7',
        '±': 'U+00B1',
        '\'': 'U+0027',
        '"': 'U+0022',
        ':': 'U+003A',
        ';': 'U+003B',
        '`': 'U+0060',
        '~': 'U+007E',
    },
};
