// testcafe-hammerhead/lib/utils/promisified-functions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.exec = exports.fsObjectExists = exports.writeFile = exports.makeDir = exports.access = exports.stat = exports.readFile = exports.readDir = exports.brotliDecompress = exports.brotliCompress = exports.inflateRaw = exports.inflate = exports.gunzip = exports.deflate = exports.gzip = void 0;
const zlib_1 = __importDefault(require("zlib"));
const util_1 = require("util");
const fs_1 = __importDefault(require("fs"));
const child_process_1 = __importDefault(require("child_process"));
exports.gzip = util_1.promisify(zlib_1.default.gzip);
exports.deflate = util_1.promisify(zlib_1.default.deflate);
exports.gunzip = util_1.promisify(zlib_1.default.gunzip);
exports.inflate = util_1.promisify(zlib_1.default.inflate);
exports.inflateRaw = util_1.promisify(zlib_1.default.inflateRaw);
exports.brotliCompress = util_1.promisify(zlib_1.default.brotliCompress);
exports.brotliDecompress = util_1.promisify(zlib_1.default.brotliDecompress);
exports.readDir = util_1.promisify(fs_1.default.readdir);
exports.readFile = util_1.promisify(fs_1.default.readFile);
exports.stat = util_1.promisify(fs_1.default.stat);
exports.access = util_1.promisify(fs_1.default.access);
exports.makeDir = util_1.promisify(fs_1.default.mkdir);
exports.writeFile = util_1.promisify(fs_1.default.writeFile);
exports.fsObjectExists = (fsPath) => exports.stat(fsPath).then(() => true, () => false);
exports.exec = util_1.promisify(child_process_1.default.exec);
