// testcafe/lib/test-run/commands/validations/factories.js
"use strict";
// -------------------------------------------------------------
// WARNING: this file is used by both the client and the server.
// Do not use any browser or node-specific API!
// -------------------------------------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFunctionValidator = exports.createObjectValidator = exports.createUrlSearchParamsValidator = exports.createUrlValidator = exports.createNumberValidator = exports.createDateValidator = exports.createStringOrRegexValidator = exports.createStringValidator = exports.createSpeedValidator = exports.createBooleanValidator = exports.createPositiveIntegerValidator = exports.createIntegerValidator = void 0;
function createIntegerValidator(ErrorCtor) {
    return (name, val) => {
        const valType = typeof val;
        if (valType !== 'number')
            throw new ErrorCtor(name, valType);
        const isInteger = !isNaN(val) &&
            isFinite(val) &&
            val === Math.floor(val);
        if (!isInteger)
            throw new ErrorCtor(name, val);
    };
}
exports.createIntegerValidator = createIntegerValidator;
function createPositiveIntegerValidator(ErrorCtor) {
    const integerValidator = createIntegerValidator(ErrorCtor);
    return (name, val) => {
        integerValidator(name, val);
        if (val < 0)
            throw new ErrorCtor(name, val);
    };
}
exports.createPositiveIntegerValidator = createPositiveIntegerValidator;
function createBooleanValidator(ErrorCtor) {
    return (name, val) => {
        const valType = typeof val;
        if (valType !== 'boolean')
            throw new ErrorCtor(name, valType);
    };
}
exports.createBooleanValidator = createBooleanValidator;
function createSpeedValidator(ErrorCtor) {
    return (name, val) => {
        const valType = typeof val;
        if (valType !== 'number')
            throw new ErrorCtor(name, valType);
        if (isNaN(val) || val < 0.01 || val > 1)
            throw new ErrorCtor(name, val);
    };
}
exports.createSpeedValidator = createSpeedValidator;
function createStringValidator(ErrorCtor) {
    return (name, val) => {
        const valType = typeof val;
        if (valType !== 'string')
            throw new ErrorCtor(name, valType);
    };
}
exports.createStringValidator = createStringValidator;
function createStringOrRegexValidator(ErrorCtor) {
    return (name, val) => {
        const valType = typeof val;
        if (valType !== 'string' && !(val instanceof RegExp))
            throw new ErrorCtor(name, valType);
    };
}
exports.createStringOrRegexValidator = createStringOrRegexValidator;
function createDateValidator(ErrorCtor) {
    return (name, val) => {
        if (!(val instanceof Date))
            throw new ErrorCtor(name, val);
    };
}
exports.createDateValidator = createDateValidator;
function createNumberValidator(ErrorCtor) {
    return (name, val) => {
        if (isNaN(Number(val)))
            throw new ErrorCtor(name, typeof val);
    };
}
exports.createNumberValidator = createNumberValidator;
function createUrlValidator(ErrorCtor) {
    return (name, val) => {
        const valType = typeof val;
        if (valType !== 'string' && !(val instanceof URL))
            throw new ErrorCtor(name, valType);
    };
}
exports.createUrlValidator = createUrlValidator;
function createUrlSearchParamsValidator(ErrorCtor) {
    return (name, val) => {
        const valType = typeof val;
        if (valType !== 'object' && !(val instanceof URLSearchParams))
            throw new ErrorCtor(name, valType);
    };
}
exports.createUrlSearchParamsValidator = createUrlSearchParamsValidator;
function createObjectValidator(ErrorCtor) {
    return (name, val) => {
        const valType = typeof val;
        if (valType !== 'object')
            throw new ErrorCtor(name, valType);
    };
}
exports.createObjectValidator = createObjectValidator;
function createFunctionValidator(ErrorCtor) {
    return (name, val) => {
        const valType = typeof val;
        if (valType !== 'function')
            throw new ErrorCtor(name, valType);
    };
}
exports.createFunctionValidator = createFunctionValidator;
