// testcafe/lib/services/serialization/replicator/transforms/request-hook-event-data-transform.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const testcafe_hammerhead_1 = require("testcafe-hammerhead");
const EVENT_CONSTRUCTORS = [
    testcafe_hammerhead_1.RequestEvent,
    testcafe_hammerhead_1.ConfigureResponseEvent,
    testcafe_hammerhead_1.ResponseEvent,
];
const EVENT_CONSTRUCTORS_MAP = new Map(EVENT_CONSTRUCTORS.map(eventConstructor => {
    return [eventConstructor.name, eventConstructor];
}));
class RequestHookEventDataTransform {
    constructor() {
        this.type = 'RequestHookEventData';
    }
    shouldTransform(_, val) {
        return EVENT_CONSTRUCTORS.some(EventConstructor => val instanceof EventConstructor);
    }
    toSerializable(value) {
        // TODO: Remove eventData._requestContext into 'testcafe-hammerhead' module
        // after switching to the compiler service mode.
        // NOTE: Access to the deprecated property inside of the unserializable 'eventData._requestContext' property
        // causes the node's deprecation warning.
        const clonedEventData = Object.assign({}, value);
        // @ts-ignore
        delete clonedEventData._requestContext;
        return {
            constructorName: value.constructor.name,
            data: clonedEventData,
        };
    }
    fromSerializable(value) {
        const EventConstructor = EVENT_CONSTRUCTORS_MAP.get(value.constructorName);
        if (!EventConstructor)
            throw new Error(`An appropriate command constructor for "${value.constructorName}" type was not found.`);
        return EventConstructor.from(value.data);
    }
}
exports.default = RequestHookEventDataTransform;
module.exports = exports.default;
