// testcafe-hammerhead/lib/request-pipeline/request-hooks/events/factory/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base_1 = __importDefault(require("./base"));
const info_1 = require("../info");
const request_options_1 = __importDefault(require("../../../request-options"));
const configure_response_event_1 = __importDefault(require("../configure-response-event"));
class RequestPipelineRequestHookEventFactory extends base_1.default {
    constructor(ctx) {
        super();
        this._ctx = ctx;
    }
    createRequestInfo() {
        return info_1.RequestInfo.from(this._ctx);
    }
    createRequestOptions() {
        return request_options_1.default.createFrom(this._ctx);
    }
    createConfigureResponseEvent(rule) {
        return new configure_response_event_1.default(rule, {
            setHeader: (name, value) => {
                this._ctx.destRes.headers[name.toLowerCase()] = value;
            },
            removeHeader: name => {
                delete this._ctx.destRes.headers[name.toLowerCase()];
            },
        });
    }
    createResponseInfo() {
        return info_1.ResponseInfo.from(this._ctx);
    }
}
exports.default = RequestPipelineRequestHookEventFactory;module.exports = exports.default;
