// testcafe-hammerhead/lib/request-pipeline/request-hooks/response-mock/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const generate_unique_id_1 = __importDefault(require("../../../utils/generate-unique-id"));
const validate_parameters_1 = __importDefault(require("./validate-parameters"));
const lowercase_header_names_1 = __importDefault(require("./lowercase-header-names"));
const is_predicate_1 = __importDefault(require("../is-predicate"));
// NOTE: ResponseMock is a data transfer object
// It should contain only initialization and creation logic
class ResponseMock {
    constructor(body, statusCode, headers) {
        validate_parameters_1.default(body, statusCode, headers);
        this.id = generate_unique_id_1.default();
        this.body = body;
        this.isPredicate = is_predicate_1.default(body);
        if (headers)
            this.headers = lowercase_header_names_1.default(headers);
        if (statusCode)
            this.statusCode = statusCode;
    }
    setRequestOptions(opts) {
        this.requestOptions = opts;
    }
    static from(val) {
        if (!val)
            return null;
        const body = val['body'];
        const statusCode = val['statusCode'];
        const headers = val['headers'];
        return new ResponseMock(body, statusCode, headers);
    }
}
exports.default = ResponseMock;module.exports = exports.default;
