// testcafe/lib/configuration/option-names.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var OptionNames;
(function (OptionNames) {
    OptionNames["src"] = "src";
    OptionNames["browsers"] = "browsers";
    OptionNames["concurrency"] = "concurrency";
    OptionNames["filter"] = "filter";
    OptionNames["filterTestGrep"] = "filter.testGrep";
    OptionNames["filterFixtureGrep"] = "filter.fixtureGrep";
    OptionNames["reporter"] = "reporter";
    OptionNames["ssl"] = "ssl";
    OptionNames["speed"] = "speed";
    OptionNames["screenshots"] = "screenshots";
    OptionNames["screenshotPath"] = "screenshotPath";
    OptionNames["screenshotPathPattern"] = "screenshotPathPattern";
    OptionNames["screenshotPathPatternOnFails"] = "screenshotPathPatternOnFails";
    OptionNames["takeScreenshotsOnFails"] = "takeScreenshotsOnFails";
    OptionNames["proxyBypass"] = "proxyBypass";
    OptionNames["appCommand"] = "appCommand";
    OptionNames["appInitDelay"] = "appInitDelay";
    OptionNames["skipJsErrors"] = "skipJsErrors";
    OptionNames["quarantineMode"] = "quarantineMode";
    OptionNames["debugMode"] = "debugMode";
    OptionNames["debugOnFail"] = "debugOnFail";
    OptionNames["skipUncaughtErrors"] = "skipUncaughtErrors";
    OptionNames["stopOnFirstFail"] = "stopOnFirstFail";
    OptionNames["selectorTimeout"] = "selectorTimeout";
    OptionNames["assertionTimeout"] = "assertionTimeout";
    OptionNames["pageLoadTimeout"] = "pageLoadTimeout";
    OptionNames["browserInitTimeout"] = "browserInitTimeout";
    OptionNames["testExecutionTimeout"] = "testExecutionTimeout";
    OptionNames["runExecutionTimeout"] = "runExecutionTimeout";
    OptionNames["videoPath"] = "videoPath";
    OptionNames["videoOptions"] = "videoOptions";
    OptionNames["videoEncodingOptions"] = "videoEncodingOptions";
    OptionNames["tsConfigPath"] = "tsConfigPath";
    OptionNames["clientScripts"] = "clientScripts";
    OptionNames["requestHooks"] = "requestHooks";
    OptionNames["retryTestPages"] = "retryTestPages";
    OptionNames["hostname"] = "hostname";
    OptionNames["port1"] = "port1";
    OptionNames["port2"] = "port2";
    OptionNames["developmentMode"] = "developmentMode";
    OptionNames["disablePageReloads"] = "disablePageReloads";
    OptionNames["disablePageCaching"] = "disablePageCaching";
    OptionNames["disableScreenshots"] = "disableScreenshots";
    OptionNames["debugLogger"] = "debugLogger";
    OptionNames["disableMultipleWindows"] = "disableMultipleWindows";
    OptionNames["experimentalMultipleWindows"] = "experimentalMultipleWindows";
    OptionNames["disableHttp2"] = "disableHttp2";
    OptionNames["disableNativeAutomation"] = "disableNativeAutomation";
    OptionNames["compilerOptions"] = "compilerOptions";
    OptionNames["pageRequestTimeout"] = "pageRequestTimeout";
    OptionNames["ajaxRequestTimeout"] = "ajaxRequestTimeout";
    OptionNames["cache"] = "cache";
    OptionNames["userVariables"] = "userVariables";
    OptionNames["v8Flags"] = "v8Flags";
    OptionNames["hooks"] = "hooks";
    OptionNames["baseUrl"] = "baseUrl";
    OptionNames["disableCrossDomain"] = "disableCrossDomain";
    OptionNames["esm"] = "esm";
    OptionNames["customActions"] = "customActions";
})(OptionNames || (OptionNames = {}));
exports.default = OptionNames;
module.exports = exports.default;
