// testcafe/lib/configuration/skip-js-errors-option-names.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SKIP_JS_ERRORS_CALLBACK_WITH_OPTIONS_OPTION_NAMES = exports.SKIP_JS_ERRORS_OPTIONS_OBJECT_OPTION_NAMES = void 0;
var SKIP_JS_ERRORS_OPTIONS_OBJECT_OPTION_NAMES;
(function (SKIP_JS_ERRORS_OPTIONS_OBJECT_OPTION_NAMES) {
    SKIP_JS_ERRORS_OPTIONS_OBJECT_OPTION_NAMES["message"] = "message";
    SKIP_JS_ERRORS_OPTIONS_OBJECT_OPTION_NAMES["stack"] = "stack";
    SKIP_JS_ERRORS_OPTIONS_OBJECT_OPTION_NAMES["pageUrl"] = "pageUrl";
})(SKIP_JS_ERRORS_OPTIONS_OBJECT_OPTION_NAMES = exports.SKIP_JS_ERRORS_OPTIONS_OBJECT_OPTION_NAMES || (exports.SKIP_JS_ERRORS_OPTIONS_OBJECT_OPTION_NAMES = {}));
var SKIP_JS_ERRORS_CALLBACK_WITH_OPTIONS_OPTION_NAMES;
(function (SKIP_JS_ERRORS_CALLBACK_WITH_OPTIONS_OPTION_NAMES) {
    SKIP_JS_ERRORS_CALLBACK_WITH_OPTIONS_OPTION_NAMES["fn"] = "fn";
    SKIP_JS_ERRORS_CALLBACK_WITH_OPTIONS_OPTION_NAMES["dependencies"] = "dependencies";
})(SKIP_JS_ERRORS_CALLBACK_WITH_OPTIONS_OPTION_NAMES = exports.SKIP_JS_ERRORS_CALLBACK_WITH_OPTIONS_OPTION_NAMES || (exports.SKIP_JS_ERRORS_CALLBACK_WITH_OPTIONS_OPTION_NAMES = {}));
