// testcafe/lib/browser/provider/utils/argument-parsing.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseConfig = exports.getModes = exports.getPathFromParsedModes = exports.splitEscaped = exports.isMatchTrue = exports.findMatch = exports.hasMatch = void 0;
const lodash_1 = require("lodash");
const os_family_1 = __importDefault(require("os-family"));
const CONFIG_TERMINATOR_RE = /(\s+|^)-/;
function hasMatch(array, re) {
    return !!(0, lodash_1.find)(array, el => el.match(re));
}
exports.hasMatch = hasMatch;
function findMatch(array, re) {
    const element = (0, lodash_1.find)(array, el => el.match(re));
    return element ? element.match(re)[1] : '';
}
exports.findMatch = findMatch;
function isMatchTrue(array, re) {
    const match = findMatch(array, re);
    return match && match !== '0' && match !== 'false';
}
exports.isMatchTrue = isMatchTrue;
function splitEscaped(str, splitterChar) {
    const result = [''];
    for (let i = 0; i < str.length; i++) {
        if (str[i] === splitterChar) {
            result.push('');
            continue;
        }
        if (str[i] === '\\' && (str[i + 1] === '\\' || str[i + 1] === splitterChar))
            i++;
        result[result.length - 1] += str[i];
    }
    return result;
}
exports.splitEscaped = splitEscaped;
function getPathFromParsedModes(modes, availableModes = []) {
    if (!modes.length)
        return '';
    if (availableModes.some(mode => mode === modes[0]))
        return '';
    let path = modes.shift();
    if (os_family_1.default.win && modes.length && path.match(/^[A-Za-z]$/))
        path += ':' + modes.shift();
    return path;
}
exports.getPathFromParsedModes = getPathFromParsedModes;
function getModes(modes, availableModes = []) {
    const result = {};
    availableModes = availableModes.slice();
    availableModes.forEach(key => {
        result[key] = false;
    });
    while (modes.length && availableModes.length) {
        if (modes[0] === availableModes[0]) {
            result[availableModes[0]] = true;
            modes.shift();
        }
        availableModes.shift();
    }
    return result;
}
exports.getModes = getModes;
function parseConfig(str) {
    const configTerminatorMatch = str.match(CONFIG_TERMINATOR_RE);
    if (!configTerminatorMatch)
        return { modesString: str, userArgs: '' };
    return {
        modesString: str.substr(0, configTerminatorMatch.index),
        userArgs: str.substr(configTerminatorMatch.index + configTerminatorMatch[1].length),
    };
}
exports.parseConfig = parseConfig;
