// testcafe/lib/browser/provider/utils/browser-starter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const os_family_1 = __importDefault(require("os-family"));
const testcafe_browser_tools_1 = __importDefault(require("testcafe-browser-tools"));
const delay_1 = __importDefault(require("../../../utils/delay"));
const POST_OPERATION_DELAY = 500;
class OperationsQueue {
    constructor() {
        this.chainPromise = Promise.resolve();
    }
    executeOperation(operation) {
        const operationPromise = this.chainPromise.then(operation);
        this.chainPromise = operationPromise.then(() => (0, delay_1.default)(POST_OPERATION_DELAY));
        return operationPromise;
    }
}
class BrowserStarter {
    constructor() {
        // NOTE: You can't start multiple instances of the same app at the same time on macOS.
        // That's why a queue of opening requests is needed.
        this.macOSBrowserOpeningQueue = new OperationsQueue();
    }
    async startBrowser(...openArgs) {
        const openBrowserOperation = () => testcafe_browser_tools_1.default.open(...openArgs);
        if (os_family_1.default.mac)
            await this.macOSBrowserOpeningQueue.executeOperation(openBrowserOperation);
        else
            await openBrowserOperation();
    }
}
exports.default = BrowserStarter;
module.exports = exports.default;
