// testcafe/lib/video-recorder/test-run-video-recorder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const process_1 = __importDefault(require("./process"));
const VIDEO_EXTENSION = 'mp4';
const TEMP_VIDEO_FILE_PREFIX = 'tmp-video';
const TEMP_MERGE_FILE_PREFIX = TEMP_VIDEO_FILE_PREFIX + '-merge';
const TEMP_MERGE_CONFIG_FILE_PREFIX = 'config';
const TEMP_MERGE_CONFIG_FILE_EXTENSION = 'txt';
class TestRunVideoRecorder {
    constructor({ testRun, test, index }, { path, ffmpegPath, encodingOptions }) {
        this.testRun = testRun;
        this.test = test;
        this.index = index;
        this.tempFiles = null;
        this.videoRecorder = null;
        this.path = path;
        this.ffmpegPath = ffmpegPath;
        this.encodingOptions = encodingOptions;
        this.start = null;
        this.timecodes = null;
    }
    get testRunInfo() {
        return {
            testIndex: this.index,
            fixture: this.test.fixture.name,
            test: this.test.name,
            timecodes: this.timecodes,
            alias: this._connection.browserInfo.alias,
            parsedUserAgent: this._connection.browserInfo.parsedUserAgent,
        };
    }
    get hasErrors() {
        return !!this.testRun.errs.length;
    }
    get _connection() {
        return this.testRun.browserConnection;
    }
    async startCapturing() {
        await this.videoRecorder.startCapturing();
        this.start = Date.now();
    }
    async finishCapturing() {
        await this.videoRecorder.finishCapturing();
    }
    async init() {
        this.tempFiles = this._generateTempNames();
        this.videoRecorder = this._createVideoRecorderProcess();
        await this.videoRecorder.init();
    }
    async isVideoSupported() {
        const connectionCapabilities = await this._connection.provider.hasCustomActionForBrowser(this._connection.id);
        return connectionCapabilities && connectionCapabilities.hasGetVideoFrameData;
    }
    async isVideoEnabled() {
        return !this.test.skip;
    }
    onTestRunRestart() {
        this.timecodes = this.timecodes || [0];
        this._addTimeCode();
    }
    _addTimeCode() {
        this.timecodes.push(Date.now() - this.start);
    }
    _createVideoRecorderProcess() {
        return new process_1.default(this.tempFiles.tempVideoPath, this.ffmpegPath, this._connection, this.encodingOptions);
    }
    _generateTempNames() {
        const id = this._connection.id;
        const tempFileNames = {
            tempVideoPath: `${TEMP_VIDEO_FILE_PREFIX}-${id}.${VIDEO_EXTENSION}`,
            tempMergeConfigPath: `${TEMP_MERGE_CONFIG_FILE_PREFIX}-${id}.${TEMP_MERGE_CONFIG_FILE_EXTENSION}`,
            tmpMergeName: `${TEMP_MERGE_FILE_PREFIX}-${id}.${VIDEO_EXTENSION}`,
        };
        for (const [tempFile, tempName] of Object.entries(tempFileNames))
            tempFileNames[tempFile] = (0, path_1.join)(this.path, tempName);
        return tempFileNames;
    }
}
exports.default = TestRunVideoRecorder;
module.exports = exports.default;
