// testcafe-hammerhead/lib/request-pipeline/request-hooks/events/configure-response-event.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const configure_response_event_options_1 = __importDefault(require("./configure-response-event-options"));
const generate_unique_id_1 = __importDefault(require("../../../utils/generate-unique-id"));
class ConfigureResponseEvent {
    constructor(requestFilterRule, modifyResponseFunctions, opts = configure_response_event_options_1.default.DEFAULT) {
        this.requestFilterRule = requestFilterRule;
        this._modifyResponseFunctions = modifyResponseFunctions;
        this.opts = opts;
        this.id = (0, generate_unique_id_1.default)();
    }
    async setHeader(name, value) {
        if (!this._modifyResponseFunctions)
            return;
        this._modifyResponseFunctions.setHeader(name, value);
    }
    async removeHeader(name) {
        if (!this._modifyResponseFunctions)
            return;
        this._modifyResponseFunctions.removeHeader(name);
    }
}
exports.default = ConfigureResponseEvent;module.exports = exports.default;
