// testcafe/lib/shared/errors/element-hidden-reasons.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    notElementOrTextNode: (targetType) => `
        The ${targetType} is neither a DOM element nor a text node.
    `,
    elOutsideBounds: (target, targetType) => `
        The ${targetType} (${target}) is located outside the the layout viewport.
    `,
    elHasWidthOrHeightZero: (target, targetType, width, height) => `
        The ${targetType} (${target}) is too small to be visible: ${width}px x ${height}px.
    `,
    elHasDisplayNone: (target, targetType) => `
        The ${targetType} (${target}) is invisible. 
        The value of its 'display' property is 'none'.
    `,
    parentHasDisplayNone: (target, targetType, parent) => `
        The ${targetType} (${target}) is invisible. 
        It descends from an element that has the 'display: none' property (${parent}).
    `,
    elHasVisibilityHidden: (target, targetType) => `
        The ${targetType} (${target}) is invisible.
        The value of its 'visibility' property is 'hidden'.
    `,
    parentHasVisibilityHidden: (target, targetType, parent) => `
        The ${targetType} (${target}) is invisible.
        It descends from an element that has the 'visibility: hidden' property (${parent}).
    `,
    elHasVisibilityCollapse: (target, targetType) => `
        The ${targetType} (${target}) is invisible.
        The value of its 'visibility' property is 'collapse'.
    `,
    parentHasVisibilityCollapse: (target, targetType, parent) => `
        The ${targetType} (${target}) is invisible.
        It descends from an element that has the 'visibility: collapse' property (${parent}).
    `,
    elNotRendered: (target, targetType) => `
        The ${targetType} (${target}) has not been rendered.
    `,
    optionNotVisible: (target, targetType, parent) => `
        The ${targetType} (${target}) is invisible. 
        The parent element (${parent}) is collapsed, and its length is shorter than 2.
    `,
    mapContainerNotVisible: (target, containerHiddenReason) => `
        The action target (${target}) is invisible because ${containerHiddenReason}
    `,
};
module.exports = exports.default;
