// testcafe/lib/browser/provider/built-in/dedicated/chrome/cdp-client/utils/cursor.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CursorUICdp = exports.CursorState = void 0;
const clientsManager = __importStar(require("../clients-manager"));
const axis_values_1 = __importDefault(require("../../../../../../../shared/utils/values/axis-values"));
const DEFAULT_COLOR = { r: 50, g: 50, b: 50, a: 0.5 };
const LEFT_BUTTON_DOWN_COLOR = { r: 200, g: 50, b: 50, a: 0.5 };
const RIGHT_BUTTON_DOWN_COLOR = { r: 50, g: 200, b: 50, a: 0.5 };
const CURSOR_RECT = {
    width: 20,
    height: 20,
};
const HIGHTLIGHT_CONFIG = {
    showInfo: true,
    showStyles: true,
    showRulers: true,
    showExtensionLines: true,
    eventTargetColor: DEFAULT_COLOR,
};
var CursorState;
(function (CursorState) {
    CursorState[CursorState["default"] = 0] = "default";
    CursorState[CursorState["leftButtonDown"] = 1] = "leftButtonDown";
    CursorState[CursorState["rightButtonDown"] = 2] = "rightButtonDown";
})(CursorState = exports.CursorState || (exports.CursorState = {}));
class CursorUICdp {
    constructor() {
        this._position = new axis_values_1.default(0, 0);
        this._visible = false;
        this._state = CursorState.default;
    }
    get currentColor() {
        if (this._state === CursorState.leftButtonDown)
            return LEFT_BUTTON_DOWN_COLOR;
        if (this._state === CursorState.rightButtonDown)
            return RIGHT_BUTTON_DOWN_COLOR;
        return DEFAULT_COLOR;
    }
    get overlay() {
        return clientsManager.getClient().Overlay;
    }
    get DOM() {
        return clientsManager.getClient().DOM;
    }
    isVisible() {
        return this._visible;
    }
    async move(position) {
        this._position = position;
        if (this.isVisible())
            await this.show();
    }
    async hide() {
        this._visible = false;
        await this.overlay.hideHighlight();
    }
    async show() {
        this._visible = true;
        if (this._state === CursorState.default)
            await this.overlay.highlightRect(Object.assign({}, CURSOR_RECT, { color: this.currentColor }, this._position));
        else {
            const nodeForLocation = await this.DOM.getNodeForLocation(this._position);
            const config = { highlightConfig: Object.assign({ contentColor: this.currentColor }, HIGHTLIGHT_CONFIG) };
            await this.overlay.highlightNode(Object.assign(nodeForLocation, config));
        }
    }
    async leftButtonDown() {
        this._state = CursorState.leftButtonDown;
        await this.show();
    }
    async rightButtonDown() {
        this._state = CursorState.rightButtonDown;
        await this.show();
    }
    async buttonUp() {
        this._state = CursorState.default;
        await this.show();
    }
}
exports.CursorUICdp = CursorUICdp;
