// testcafe-reporter-dashboard/lib/types/internal/dashboard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UploadStatus = exports.createDashboardTestRunInfo = exports.createTestError = exports.AggregateNames = exports.AggregateCommandType = void 0;
var AggregateCommandType;
(function (AggregateCommandType) {
    AggregateCommandType["reportTaskStart"] = "reportTaskStart";
    AggregateCommandType["reportTestStart"] = "reportTestStart";
    AggregateCommandType["reportTestDone"] = "reportTestDone";
    AggregateCommandType["reportTaskDone"] = "reportTaskDone";
    AggregateCommandType["createUpload"] = "createUpload";
    AggregateCommandType["reportWarnings"] = "reportWarnings";
})(AggregateCommandType = exports.AggregateCommandType || (exports.AggregateCommandType = {}));
;
var AggregateNames;
(function (AggregateNames) {
    AggregateNames["Run"] = "Run";
    AggregateNames["Upload"] = "Upload";
})(AggregateNames = exports.AggregateNames || (exports.AggregateNames = {}));
;
exports.createTestError = (error, errorModel) => ({
    code: error.code,
    testRunPhase: error.testRunPhase,
    errorModel: errorModel
});
exports.createDashboardTestRunInfo = ({ quarantine, unstable, warnings }, browserRuns) => ({
    quarantine,
    unstable,
    warnings,
    browserRuns
});
var UploadStatus;
(function (UploadStatus) {
    UploadStatus["Completed"] = "Completed";
    UploadStatus["Failed"] = "Failed";
})(UploadStatus = exports.UploadStatus || (exports.UploadStatus = {}));
;
;
