// testcafe-hammerhead/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const proxy_1 = __importDefault(require("./proxy"));
const session_1 = __importDefault(require("./session"));
const response_mock_1 = __importDefault(require("./request-pipeline/request-hooks/response-mock"));
const request_filter_rule_1 = __importDefault(require("./request-pipeline/request-hooks/request-filter-rule"));
const storage_1 = __importDefault(require("./upload/storage"));
const script_1 = require("./processing/script");
const configure_response_event_options_1 = __importDefault(require("./session/events/configure-response-event-options"));
const state_snapshot_1 = __importDefault(require("./session/state-snapshot"));
const url_1 = require("./utils/url");
const generate_unique_id_1 = __importDefault(require("./utils/generate-unique-id"));
const responseMockSetBodyMethod = __importStar(require("./request-pipeline/request-hooks/response-mock/set-body-method"));
const request_event_1 = __importDefault(require("./session/events/request-event"));
const response_event_1 = __importDefault(require("./session/events/response-event"));
const configure_response_event_1 = __importDefault(require("./session/events/configure-response-event"));
exports.default = {
    Proxy: proxy_1.default,
    Session: session_1.default,
    UploadStorage: storage_1.default,
    processScript: script_1.processScript,
    isScriptProcessed: script_1.isScriptProcessed,
    ResponseMock: response_mock_1.default,
    RequestFilterRule: request_filter_rule_1.default,
    ConfigureResponseEventOptions: configure_response_event_options_1.default,
    StateSnapshot: state_snapshot_1.default,
    SPECIAL_BLANK_PAGE: url_1.SPECIAL_BLANK_PAGE,
    SPECIAL_ERROR_PAGE: url_1.SPECIAL_ERROR_PAGE,
    generateUniqueId: generate_unique_id_1.default,
    responseMockSetBodyMethod,
    RequestEvent: request_event_1.default,
    ConfigureResponseEvent: configure_response_event_1.default,
    ResponseEvent: response_event_1.default,
};module.exports = exports.default;
