// testcafe-hammerhead/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const proxy_1 = __importDefault(require("./proxy"));
const session_1 = __importDefault(require("./session"));
const response_mock_1 = __importDefault(require("./request-pipeline/request-hooks/response-mock"));
const request_filter_rule_1 = __importDefault(require("./request-pipeline/request-hooks/request-filter-rule"));
const storage_1 = __importDefault(require("./upload/storage"));
const index_1 = require("./upload/index");
const script_1 = require("./processing/script");
const configure_response_event_options_1 = __importDefault(require("./request-pipeline/request-hooks/events/configure-response-event-options"));
const state_snapshot_1 = __importDefault(require("./session/state-snapshot"));
const url_1 = require("./utils/url");
const generate_unique_id_1 = __importDefault(require("./utils/generate-unique-id"));
const responseMockSetBodyMethod = __importStar(require("./request-pipeline/request-hooks/response-mock/set-body-method"));
const request_event_1 = __importDefault(require("./request-pipeline/request-hooks/events/request-event"));
const response_event_1 = __importDefault(require("./request-pipeline/request-hooks/events/response-event"));
const configure_response_event_1 = __importDefault(require("./request-pipeline/request-hooks/events/configure-response-event"));
const destination_request_1 = __importDefault(require("./request-pipeline/destination-request"));
const request_options_1 = __importDefault(require("./request-pipeline/request-options"));
const promisify_stream_1 = __importDefault(require("./utils/promisify-stream"));
const page_1 = __importDefault(require("./processing/resources/page"));
const injectables_1 = require("./session/injectables");
const http_1 = require("./utils/http");
const get_asset_path_1 = __importDefault(require("./utils/get-asset-path"));
const event_provider_1 = __importDefault(require("./request-pipeline/request-hooks/events/event-provider"));
const info_1 = require("./request-pipeline/request-hooks/events/info");
const base_1 = __importDefault(require("./request-pipeline/request-hooks/events/factory/base"));
const base_2 = __importDefault(require("./request-pipeline/context/base"));
const is_redirect_status_code_1 = __importDefault(require("./utils/is-redirect-status-code"));
const contentTypeUtils = __importStar(require("./utils/content-type"));
const encoding_1 = require("./utils/encoding");
exports.default = {
    Proxy: proxy_1.default,
    Session: session_1.default,
    UploadStorage: storage_1.default,
    processScript: script_1.processScript,
    isScriptProcessed: script_1.isScriptProcessed,
    ResponseMock: response_mock_1.default,
    RequestFilterRule: request_filter_rule_1.default,
    ConfigureResponseEventOptions: configure_response_event_options_1.default,
    StateSnapshot: state_snapshot_1.default,
    SPECIAL_BLANK_PAGE: url_1.SPECIAL_BLANK_PAGE,
    SPECIAL_ERROR_PAGE: url_1.SPECIAL_ERROR_PAGE,
    generateUniqueId: generate_unique_id_1.default,
    responseMockSetBodyMethod,
    RequestEvent: request_event_1.default,
    ConfigureResponseEvent: configure_response_event_1.default,
    ResponseEvent: response_event_1.default,
    DestinationRequest: destination_request_1.default,
    RequestOptions: request_options_1.default,
    RequestHookEventProvider: event_provider_1.default,
    BaseRequestHookEventFactory: base_1.default,
    BaseRequestPipelineContext: base_2.default,
    RequestInfo: info_1.RequestInfo,
    promisifyStream: promisify_stream_1.default,
    parseProxyUrl: url_1.parseProxyUrl,
    sameOriginCheck: url_1.sameOriginCheck,
    injectResources: page_1.default.injectResources,
    injectUpload: index_1.inject,
    INJECTABLE_SCRIPTS: injectables_1.SCRIPTS,
    acceptCrossOrigin: http_1.acceptCrossOrigin,
    getAssetPath: get_asset_path_1.default,
    ResponseInfo: info_1.ResponseInfo,
    isRedirectStatusCode: is_redirect_status_code_1.default,
    contentTypeUtils,
    decodeBufferToString: encoding_1.decodeBufferToString,
    encodeStringToBuffer: encoding_1.encodeStringToBuffer,
};module.exports = exports.default;
