// dotaconstants/index.ts
export { default as abilities } from './build/abilities.json';
export { default as ability_ids } from './build/ability_ids.json';
export { default as aghs_desc } from './build/aghs_desc.json';
export { default as ancients } from './build/ancients.json';
export { default as chat_wheel } from './build/chat_wheel.json';
export { default as cluster } from './build/cluster.json';
export { default as countries } from './build/countries.json';
export { default as game_mode } from './build/game_mode.json';
export { default as hero_abilities } from './build/hero_abilities.json';
export { default as hero_lore } from './build/hero_lore.json';
export { default as heroes } from './build/heroes.json';
export { default as item_colors } from './build/item_colors.json';
export { default as item_ids } from './build/item_ids.json';
export { default as items } from './build/items.json';
export { default as lobby_type } from './build/lobby_type.json';
export { default as neutral_abilities } from './build/neutral_abilities.json';
export { default as order_types } from './build/order_types.json';
export { default as patch } from './build/patch.json';
export { default as patchnotes } from './build/patchnotes.json';
export { default as permanent_buffs } from './build/permanent_buffs.json';
export { default as player_colors } from './build/player_colors.json';
export { default as region } from './build/region.json';
export { default as skillshots } from './build/skillshots.json';
export { default as xp_level } from './build/xp_level.json';
