// dotaconstants/index.ts
export { default as abilities } from './build/abilities.json' with { type: 'json' };
export { default as ability_ids } from './build/ability_ids.json' with { type: 'json' };
export { default as aghs_desc } from './build/aghs_desc.json' with { type: 'json' };
export { default as ancients } from './build/ancients.json' with { type: 'json' };
export { default as chat_wheel } from './build/chat_wheel.json' with { type: 'json' };
export { default as cluster } from './build/cluster.json' with { type: 'json' };
export { default as countries } from './build/countries.json' with { type: 'json' };
export { default as game_mode } from './build/game_mode.json' with { type: 'json' };
export { default as hero_abilities } from './build/hero_abilities.json' with { type: 'json' };
export { default as hero_lore } from './build/hero_lore.json' with { type: 'json' };
export { default as heroes } from './build/heroes.json' with { type: 'json' };
export { default as item_colors } from './build/item_colors.json' with { type: 'json' };
export { default as item_ids } from './build/item_ids.json' with { type: 'json' };
export { default as items } from './build/items.json' with { type: 'json' };
export { default as lobby_type } from './build/lobby_type.json' with { type: 'json' };
export { default as neutral_abilities } from './build/neutral_abilities.json' with { type: 'json' };
export { default as order_types } from './build/order_types.json' with { type: 'json' };
export { default as patch } from './build/patch.json' with { type: 'json' };
export { default as patchnotes } from './build/patchnotes.json' with { type: 'json' };
export { default as permanent_buffs } from './build/permanent_buffs.json' with { type: 'json' };
export { default as player_colors } from './build/player_colors.json' with { type: 'json' };
export { default as region } from './build/region.json' with { type: 'json' };
export { default as skillshots } from './build/skillshots.json' with { type: 'json' };
export { default as xp_level } from './build/xp_level.json' with { type: 'json' };
export * as default from './index.js';
