// testcafe/lib/native-automation/client/input.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("../types");
const event_descriptor_1 = __importDefault(require("./event-descriptor"));
const hammerhead_1 = __importDefault(require("./deps/hammerhead"));
const TOUCH_MODE = hammerhead_1.default.utils.featureDetection.isTouchDevice;
class NativeAutomationInput {
    constructor(dispatchEventFn) {
        this._dispatchEventFn = dispatchEventFn;
    }
    async _executeTouchEvent(touchEvent, options) {
        const eventOptions = await event_descriptor_1.default.createTouchEventOptions(touchEvent, options);
        return this._dispatchEventFn.single(types_1.EventType.Touch, eventOptions);
    }
    async _executeMouseEvent(mouseEvent, options) {
        const eventOptions = await event_descriptor_1.default.createMouseEventOptions(mouseEvent, options);
        return this._dispatchEventFn.single(types_1.EventType.Mouse, eventOptions);
    }
    async mouseDown(options) {
        if (TOUCH_MODE)
            return this._executeTouchEvent('touchStart', options);
        return this._executeMouseEvent('mousePressed', options);
    }
    async mouseUp(options) {
        if (TOUCH_MODE)
            return this._executeTouchEvent('touchEnd', options);
        return this._executeMouseEvent('mouseReleased', options);
    }
    keyDown(options) {
        const eventOptions = event_descriptor_1.default.createKeyDownOptions(options);
        return this._dispatchEventFn.single(types_1.EventType.Keyboard, eventOptions);
    }
    keyUp(options) {
        const eventOptions = event_descriptor_1.default.createKeyUpOptions(options);
        return this._dispatchEventFn.single(types_1.EventType.Keyboard, eventOptions);
    }
    executeEventSequence(eventSequence) {
        return this._dispatchEventFn.sequence(eventSequence);
    }
    async executeInsertText(text) {
        return this._dispatchEventFn.single(types_1.EventType.InsertText, { text });
    }
    async createMouseMoveEvent(currPosition, modifiers) {
        const options = await event_descriptor_1.default.createMouseEventOptions('mouseMoved', {
            options: Object.assign({ clientX: currPosition.x, clientY: currPosition.y, button: 'none' }, modifiers),
        });
        return {
            type: types_1.EventType.Mouse,
            options,
        };
    }
}
exports.default = NativeAutomationInput;
module.exports = exports.default;
