// testcafe/lib/compiler/babel/load-libs.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const exportble_lib_path_1 = __importDefault(require("../test-file/exportble-lib-path"));
function getPresetEnvForTestCodeOpts(isCompilerServiceMode) {
    const opts = {
        targets: { node: 'current' },
        loose: true,
        exclude: ['transform-regenerator'],
    };
    if (isCompilerServiceMode)
        opts.modules = false;
    return opts;
}
function getPresetEnvForClientFunctionOpts() {
    return {
        loose: true,
        exclude: ['transform-typeof-symbol', 'transform-for-of'],
    };
}
function getModuleResolverOpts() {
    return {
        resolvePath(source) {
            if (source === 'testcafe')
                return exportble_lib_path_1.default;
            return source;
        },
    };
}
function getTransformForOfOptions() {
    // NOTE: allowArrayLike is required to allow iterating non-iterable objects (e.g. NodeList)
    // to preserve compatibility with older TestCafe code
    return { loose: true, allowArrayLike: true };
}
function getTransformRuntimeOpts() {
    // NOTE: We are forced to import helpers to each compiled file
    // because of '@babel/plugin-transform-runtime' plugin cannot correctly resolve path
    // to the helpers from the '@babel/runtime' module.
    return {
        'helpers': false,
    };
}
function getPresetReact() {
    const presetReact = require('@babel/preset-react');
    presetReact.presets = []; // disables flow so it doesn't confict w/ presetFlow
    return presetReact;
}
// NOTE: lazy load heavy dependencies
function loadLibs(isCompilerServiceMode) {
    return {
        babel: require('@babel/core'),
        presetStage2: require('./preset-stage-2'),
        presetFlow: require('@babel/preset-flow'),
        transformRuntime: [require('@babel/plugin-transform-runtime'), getTransformRuntimeOpts()],
        transformForOfAsArray: [require('@babel/plugin-transform-for-of'), getTransformForOfOptions()],
        presetEnvForClientFunction: [require('@babel/preset-env'), getPresetEnvForClientFunctionOpts()],
        presetEnvForTestCode: [require('@babel/preset-env'), getPresetEnvForTestCodeOpts(isCompilerServiceMode)],
        moduleResolver: [require('babel-plugin-module-resolver'), getModuleResolverOpts()],
        presetReact: getPresetReact(),
        proposalPrivateMethods: [require('@babel/plugin-proposal-private-methods'), { loose: true }],
        proposalClassProperties: [require('@babel/plugin-proposal-class-properties'), { loose: true }],
    };
}
exports.default = loadLibs;
module.exports = exports.default;
